--- a/drivers/usb/serial/option.c
+++ b/drivers/usb/serial/option.c
@@ -2341,6 +2341,8 @@ static const struct usb_device_id option
 	{ USB_DEVICE_INTERFACE_CLASS(0x2cb7, 0x0a05, 0xff) },			/* Fibocom FM650-CN (NCM mode) */
 	{ USB_DEVICE_INTERFACE_CLASS(0x2cb7, 0x0a06, 0xff) },			/* Fibocom FM650-CN (RNDIS mode) */
 	{ USB_DEVICE_INTERFACE_CLASS(0x2cb7, 0x0a07, 0xff) },			/* Fibocom FM650-CN (MBIM mode) */
+	{ USB_DEVICE_INTERFACE_CLASS(0x2cb7, 0x0a0b, 0xff) },			/* Fibocom MC661-CN */
+	{ USB_DEVICE_INTERFACE_CLASS(0x2cb7, 0x0a0a, 0xff) },			/* Fibocom MC661-CN(2) */
	{ USB_DEVICE_AND_INTERFACE_INFO(0x2dee, 0x4d41, 0xff, 0, 0) },		/* MeiG Smart SLM320 */
	{ USB_DEVICE_AND_INTERFACE_INFO(0x2dee, 0x4d42, 0xff, 0, 0) },		/* MeiG Smart SLM332 */
	{ USB_DEVICE_AND_INTERFACE_INFO(0x2dee, 0x4d57, 0xff, 0, 0) },		/* MeiG Smart SLM770A */
