--- a/drivers/net/ethernet/stmicro/stmmac/stmmac_platform.c
+++ b/drivers/net/ethernet/stmicro/stmmac/stmmac_platform.c
@@ -462,6 +462,8 @@ stmmac_probe_config_dt(struct platform_d
 		if (bus_id < 0)
 			bus_id = -1;
 		plat->bus_id = ++bus_id;
+	} else {
+		dev_info(&pdev->dev, "Got alias ethernet%d node:%pOF\n", plat->bus_id, np);
 	}

 	/* Default to phy auto-detection */
--- a/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
+++ b/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
@@ -7671,6 +7671,9 @@ int stmmac_dvr_probe(struct device *devi
 		if (ret)
 			goto error_xpcs_setup;
 	}
+	if (priv->plat->bus_id > 0) {
+		sprintf(ndev->name, "eth%d", priv->plat->bus_id);
+	}

 	ret = stmmac_phy_setup(priv);
 	if (ret) {
