--- a/arch/arm/boot/dts/rockchip/rk3506.dtsi
+++ b/arch/arm/boot/dts/rockchip/rk3506.dtsi
@@ -102,6 +102,12 @@
 		};
 	};
 
+	cpuinfo {
+		compatible = "rockchip,cpuinfo";
+		nvmem-cells = <&otp_id>, <&otp_cpu_version>, <&cpu_code>;
+		nvmem-cell-names = "id", "cpu-version", "cpu-code";
+	};
+
 	psci {
 		compatible = "arm,psci-1.0";
 		method = "smc";
@@ -431,6 +437,58 @@
 		reg = <0xff4d8000 0x8000>;
 	};
 
+	otp: otp@ff4f0000 {
+		compatible = "rockchip,rk3506-otp";
+		reg = <0xff4f0000 0x4000>;
+		#address-cells = <1>;
+		#size-cells = <1>;
+		clocks = <&cru CLK_USER_OTPC_NS>, <&cru CLK_SBPI_OTPC_NS>,
+			 <&cru PCLK_OTPC_NS>;
+		clock-names = "usr", "sbpi", "apb";
+		resets = <&cru SRST_USER_OTPC_NS>, <&cru SRST_SBPI_OTPC_NS>,
+			 <&cru SRST_P_OTPC_NS>;
+		reset-names = "usr", "sbpi", "apb";
+
+		/* Data cells */
+		cpu_code: cpu-code@2 {
+			reg = <0x02 0x2>;
+		};
+		otp_cpu_version: cpu-version@5 {
+			reg = <0x05 0x1>;
+			bits = <3 3>;
+		};
+		specification_serial_number: specification-serial-number@8 {
+			reg = <0x08 0x1>;
+			bits = <0 5>;
+		};
+		otp_id: id@a {
+			reg = <0x0a 0x10>;
+		};
+		cpu_leakage: cpu-leakage@1e {
+			reg = <0x1e 0x1>;
+		};
+		log_leakage: log-leakage@1f {
+			reg = <0x1f 0x1>;
+		};
+		cpu_tsadc_trim_l: cpu-tsadc-trim-l@20 {
+			reg = <0x20 0x1>;
+		};
+		cpu_tsadc_trim_h: cpu-tsadc-trim-h@21 {
+			reg = <0x21 0x1>;
+			bits = <0 2>;
+		};
+		cpu_opp_info: cpu-opp-info@24 {
+			reg = <0x24 0x6>;
+		};
+		cpu_pvtpll: cpu-pvtpll@30 {
+			reg = <0x30 0x2>;
+		};
+		cpu_mbist_vmin: mbist-vmin@33 {
+			reg = <0x33 0x1>;
+			bits = <0 4>;
+		};
+	};
+
 	gic: interrupt-controller@ff581000 {
 		compatible = "arm,gic-400";
 		reg = <0xff581000 0x1000>,
