--- a/drivers/tty/serial/8250/8250_dw.c
+++ b/drivers/tty/serial/8250/8250_dw.c
@@ -425,7 +425,7 @@ static int dw8250_handle_irq(struct uart
 	unsigned int iir = serial_port_in(p, UART_IIR);
 	bool rx_timeout = (iir & 0x3f) == UART_IIR_RX_TIMEOUT;
 	unsigned int quirks = d->pdata->quirks;
-	unsigned int status;
+	unsigned int status, usr, rfl;
 
 	guard(uart_port_lock_check_sysrq_irqsave)(p);
 
@@ -462,12 +462,22 @@ static int dw8250_handle_irq(struct uart
 	 * This problem has only been observed so far when not in DMA mode
 	 * so we limit the workaround only to non-DMA mode.
 	 */
+#ifndef CONFIG_ARCH_ROCKCHIP
 	if (!up->dma && rx_timeout) {
 		status = serial_lsr_in(up);
 
 		if (!(status & (UART_LSR_DR | UART_LSR_BI)))
 			serial_port_in(p, UART_RX);
 	}
+#else
+	if ((iir & 0x3f) == UART_IIR_RX_TIMEOUT) {
+		usr = serial_port_in(p, d->pdata->usr_reg);
+		status = serial_lsr_in(up);
+		rfl = serial_port_in(p, UART_RFL);
+		if (!(status & (UART_LSR_DR | UART_LSR_BI)) && !(usr & 0x1) && (rfl == 0))
+			serial_port_in(p, UART_RX);
+	}
+#endif
 
 	/* Manually stop the Rx DMA transfer when acting as flow controller */
 	if (quirks & DW_UART_QUIRK_IS_DMA_FC && up->dma && up->dma->rx_running && rx_timeout) {
