--- a/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
+++ b/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
@@ -7425,6 +7425,8 @@ int stmmac_dvr_probe(struct device *devi
 		return -ENOMEM;
 
 	SET_NETDEV_DEV(ndev, device);
+	// https://patchew.org/linux/20250717090037.4097520-1-primoz.fiser@norik.com/
+	ndev->dev.of_node = device->of_node;
 
 	priv = netdev_priv(ndev);
 	priv->device = device;
