--- a/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
+++ b/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
@@ -4051,6 +4051,12 @@ static int stmmac_open(struct net_device
 	if (ret < 0)
 		goto err_dma_resources;

+	if (!is_valid_ether_addr(priv->dev->dev_addr)) {
+		eth_hw_addr_nvmem_sun8i_chipid(priv->device, priv->dev);
+		netdev_info(priv->dev, "fix device MAC address %pM\n",
+			priv->dev->dev_addr);
+	}
+
 	ret = stmmac_init_phy(dev);
 	if (ret)
 		goto err_runtime_pm;
