--- a/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
+++ b/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
@@ -1168,6 +1168,30 @@ static int stmmac_init_phy(struct net_de
 		return ret;
 	}

+	if (ret == 0 && priv->plat->need_sun8i_ephy_hack)
+	{
+		struct phy_device* phydev = priv->dev->phydev;
+		netdev_info(priv->dev, "%s: H3 phy hacks begin\n", __func__);
+		//EPHY Initial
+		phy_write(phydev, 0x1f , 0x0100); /* switch to page 1        */
+		phy_write(phydev, 0x12 , 0x4824); /* Disable APS             */
+		phy_write(phydev, 0x1f , 0x0200); /* switchto page 2         */
+		phy_write(phydev, 0x18 , 0x0000); /* PHYAFE TRX optimization */
+		phy_write(phydev, 0x1f , 0x0600); /* switchto page 6         */
+		phy_write(phydev, 0x14 , 0x708F); /* PHYAFE TX optimization  */
+		phy_write(phydev, 0x19 , 0x0000);
+		phy_write(phydev, 0x13 , 0xf000); /* PHYAFE RX optimization  */
+		phy_write(phydev, 0x15 , 0x1530);
+		phy_write(phydev, 0x1f , 0x0800); /* switch to page 8         */
+		phy_write(phydev, 0x18 , 0x00bc); /* PHYAFE TRX optimization */
+		//disable iEEE
+		phy_write(phydev, 0x1f , 0x0100); /* switchto page 1 */
+		/* reg 0x17 bit3,set 0 to disable iEEE */
+		phy_write(phydev, 0x17 , phy_read(phydev, 0x17) & (~(1<<3)));
+		phy_write(phydev, 0x1f , 0x0000); /* switch to page 0 */
+		netdev_info(priv->dev, "%s: H3 phy hacks end\n", __func__);
+	}
+
 	/* Configure phylib's copy of the LPI timer. Normally,
 	 * phylink_config.lpi_timer_default would do this, but there is a
 	 * chance that userspace could change the eee_timer setting via sysfs
--- a/include/linux/stmmac.h
+++ b/include/linux/stmmac.h
@@ -289,6 +289,7 @@ struct plat_stmmacenet_data {
 	struct reset_control *stmmac_rst;
 	struct reset_control *stmmac_ahb_rst;
 	struct stmmac_axi *axi;
+	bool need_sun8i_ephy_hack;
 	int rss_en;
 	int mac_port_sel_speed;
 	u8 vlan_fail_q;
--- a/drivers/net/ethernet/stmicro/stmmac/dwmac-sun8i.c
+++ b/drivers/net/ethernet/stmicro/stmmac/dwmac-sun8i.c
@@ -46,6 +46,7 @@
 struct emac_variant {
 	const struct reg_field *syscon_field;
 	bool soc_has_internal_phy;
+	bool soc_has_ephy_hack;
 	bool support_mii;
 	bool support_rmii;
 	bool support_rgmii;
@@ -91,6 +92,7 @@ static const struct reg_field sun8i_ccu_
 static const struct emac_variant emac_variant_h3 = {
 	.syscon_field = &sun8i_syscon_reg_field,
 	.soc_has_internal_phy = true,
+	.soc_has_ephy_hack = true,
 	.support_mii = true,
 	.support_rmii = true,
 	.support_rgmii = true,
@@ -1195,6 +1197,7 @@ static int sun8i_dwmac_probe(struct plat
 	plat_dat->rx_coe = STMMAC_RX_COE_TYPE2;
 	plat_dat->tx_coe = 1;
 	plat_dat->flags |= STMMAC_FLAG_HAS_SUN8I;
+	plat_dat->need_sun8i_ephy_hack = gmac->variant->soc_has_ephy_hack;
 	plat_dat->bsp_priv = gmac;
 	plat_dat->init = sun8i_dwmac_init;
 	plat_dat->exit = sun8i_dwmac_exit;
