--- a/drivers/net/ethernet/allwinner/sun4i-emac.c
+++ b/drivers/net/ethernet/allwinner/sun4i-emac.c
@@ -981,6 +981,8 @@ static int emac_probe(struct platform_de
 	int ret = 0;
 	const struct emac_quirks *quirks;

+	int alias_id;
+
 	ndev = alloc_etherdev(sizeof(struct emac_board_info));
 	if (!ndev) {
 		dev_err(&pdev->dev, "could not allocate device.\n");
@@ -1066,6 +1068,12 @@ static int emac_probe(struct platform_de
 		goto out_release_sram;
 	}

+	alias_id = of_alias_get_id(np, "ethernet");
+	if (alias_id >= 0) {
+		dev_info(&pdev->dev, "Got alias ethernet%d node:%pOF\n", alias_id, np);
+		sprintf(ndev->name, "eth%d", alias_id);
+	}
+
 	/* Read MAC-address from DT */
 	ret = of_get_ethdev_address(np, ndev);
 	if (ret) {
