--- a/arch/arm/boot/dts/allwinner/Makefile
+++ b/arch/arm/boot/dts/allwinner/Makefile
@@ -270,6 +270,7 @@ dtb-$(CONFIG_MACH_SUN8I) += \
 	sun8i-t113s-rp-t113.dtb \
 	sun8i-t113s-tlink-rp-t113.dtb \
 	sun8i-t113s-tlink-dly-e102.dtb \
+	sun8i-t113s-tlink-dly-e102-spinand.dtb \
 	sun8i-t113s-lctechpi-pc-t113.dtb \
 	sun8i-t113s-netcube-nagami-basic-carrier.dtb \
 	sun8i-t113s-netcube-nagami-keypad-carrier.dtb \
--- /dev/null
+++ b/arch/arm/boot/dts/allwinner/sun8i-t113s-tlink-dly-e102-spinand.dts
@@ -0,0 +1,37 @@
+// SPDX-License-Identifier: (GPL-2.0+ or MIT)
+// Copyright (C) 2024 Dirk Chang <dirk@kooiot.com>
+
+/dts-v1/;
+
+#include "sun8i-t113s-tlink-dly-e102.dts"
+
+/ {
+	model = "ThingsLink DLY-E102 (SPI-NAND)";
+	compatible = "kooiot,tlink-dly-e102-spinand", "allwinner,sun8i-t113";
+
+	chosen {
+		bootargs = "root=/dev/fit0 rootwait";
+	};
+};
+
+&spi0 {
+	status = "okay";
+};
+
+&flash {
+	partitions {
+		compatible = "fixed-partitions";
+		#address-cells = <1>;
+		#size-cells = <1>;
+		boot@0 {
+			label = "boot";
+			reg = <0x0000000 0x0100000>;
+			read-only;
+		};
+
+		ubi0@100000 {
+			label = "ubi";
+			reg = <0x0100000 0x0000000>;
+		};
+	};
+};
