--- a/drivers/net/phy/motorcomm.c
+++ b/drivers/net/phy/motorcomm.c
@@ -231,6 +231,10 @@
 #define YTPHY_MDIO_ADDRESS_CONTROL_REG		0xA005
 #define YTPHY_MACR_EN_PHY_ADDR_0		BIT(6)

+#define YT8521_LED0_CFG_REG				0xA00C
+#define YT8521_LED1_CFG_REG				0xA00D
+#define YT8521_LED2_CFG_REG				0xA00E
+
 #define YTPHY_MISC_CONFIG_REG			0xA006
 #define YTPHY_MCR_FIBER_SPEED_MASK		BIT(0)
 #define YTPHY_MCR_FIBER_1000BX			(0x1 << 0)
@@ -1672,6 +1676,8 @@ static int yt8521_config_init(struct phy
 	struct device_node *node = phydev->mdio.dev.of_node;
 	int old_page;
 	int ret = 0;
+	int i, val;
+	char buf[64];

 	old_page = phy_select_page(phydev, YT8521_RSSR_UTP_SPACE);
 	if (old_page < 0)
@@ -1699,6 +1705,18 @@ static int yt8521_config_init(struct phy
 		if (ret < 0)
 			goto err_restore_page;
 	}
+
+	for (i = 0; i < 3; ++i) {
+		/* get led configuration */
+		snprintf(buf, 64, "motorcomm,led%d", i);
+		if (!of_property_read_u32(node, buf, &val) ) {
+			phydev_info(phydev, "led%d config val: %08x node:%pOF", i, val, node);
+			ret = ytphy_write_ext(phydev, YT8521_LED0_CFG_REG + i, val);
+			if (ret < 0) {
+				phydev_err(phydev, "write led0 config failed!\n");
+			}
+		}
+	}
 err_restore_page:
 	return phy_restore_page(phydev, old_page, ret);
 }
