From d0e93157718befcc5f7a94ec60eb76e43b07d58d Mon Sep 17 00:00:00 2001
From: Icenowy Zheng <icenowy@aosc.io>
Date: Fri, 21 Feb 2020 13:56:29 +0800
Subject: [PATCH 6117/6117] fixup! s3: fix panel clk phase

Signed-off-by: Icenowy Zheng <icenowy@aosc.io>
---
 drivers/gpu/drm/panel/panel-simple.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/gpu/drm/panel/panel-simple.c b/drivers/gpu/drm/panel/panel-simple.c
index 6a493b4e18cf..d9f8f11a34e0 100644
--- a/drivers/gpu/drm/panel/panel-simple.c
+++ b/drivers/gpu/drm/panel/panel-simple.c
@@ -1360,6 +1360,7 @@ static const struct panel_desc sitronix_st7701s = {
 		.height = 95,
 	},
 	.bus_format = MEDIA_BUS_FMT_RGB666_1X18,
+	.bus_flags = DRM_BUS_FLAG_PIXDATA_DRIVE_POSEDGE,
 };
 
 static const struct display_timing hannstar_hsd070pww1_timing = {
-- 
2.30.2

