--- a/arch/arm/boot/dts/sun8i-v3s.dtsi
--- b/arch/arm/boot/dts/sun8i-v3s.dtsi
@@ -278,6 +278,31 @@
 			#size-cells = <0>;
 		};
 
+		sid: eeprom@1c23600 {
+			compatible = "allwinner,sun8i-a83t-sid";
+			reg = <0x01c23600 0x400>;
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			/* Data cells */
+			chipid: chipid@00 {
+				reg = <0x00 0x10>;
+			};
+			mac_nv1: mac_nv1@14 {
+				reg = <0x14 0x4>;
+			};
+			mac_nv2: mac_nv2@14 {
+				reg = <0x14 0x8>;
+			};
+			thermal_calibration: calib@34 {
+				reg = <0x34 0x8>;
+			};
+			product_sn: product_sn@d4 {
+				reg = <0xd4 0x18>;
+			};
+		};
+
+
 		usb_otg: usb@1c19000 {
 			compatible = "allwinner,sun8i-h3-musb";
 			reg = <0x01c19000 0x0400>;
@@ -579,6 +604,8 @@
 			phy-handle = <&int_mii_phy>;
 			phy-mode = "mii";
 			status = "disabled";
+			nvmem-cells = <&chipid>;
+			nvmem-cell-names = "chipid";

 			mdio: mdio {
 				#address-cells = <1>;

