--- a/drivers/regulator/core.c
+++ b/drivers/regulator/core.c
@@ -1133,6 +1133,9 @@ static int machine_constraints_voltage(struct regulator_dev *rdev,
 		}
 
 		if (current_uV < 0) {
+			if (current_uV == -EINVAL && rdev->supply_name)
+				return -EPROBE_DEFER; // DIRK: we got this from mail list, we will remove this pacth when dc1sw fixed in kernel mainline. 
+
 			rdev_err(rdev,
 				 "failed to get the current voltage: %pe\n",
 				 ERR_PTR(current_uV));
