--- a/drivers/gpu/drm/sun4i/sun4i_drv.c
+++ b/drivers/gpu/drm/sun4i/sun4i_drv.c
@@ -177,6 +177,13 @@
 	return of_device_is_compatible(node, "hdmi-connector");
 }
 
+static bool sun4i_drv_node_is_mipi_dsi(struct device_node *node)
+{
+	return of_device_is_compatible(node, "allwinner,sun8i-r40-mipi-dsi") ||
+		of_device_is_compatible(node, "allwinner,sun50i-a64-mipi-dsi") ||
+		of_device_is_compatible(node, "allwinner,sun6i-a31-mipi-dsi");
+}
+
 static bool sun4i_drv_node_is_frontend(struct device_node *node)
 {
 	return of_device_is_compatible(node, "allwinner,sun4i-a10-display-frontend") ||
@@ -346,6 +353,12 @@
 		return 0;
 
 	/*
+	 * MIPI-DSI will be the last nodes in our pipeline
+	 */
+	if (sun4i_drv_node_is_mipi_dsi(node))
+		return 0;
+
+	/*
 	 * If the device is either just a regular device, or an
 	 * enabled frontend supported by the driver, we add it to our
 	 * component list.
--- a/arch/arm/boot/dts/sun8i-r40.dtsi
+++ b/arch/arm/boot/dts/sun8i-r40.dtsi
@@ -1321,7 +1321,7 @@
 			compatible = "allwinner,sun8i-r40-hdmi-phy";
 			reg = <0x01ef0000 0x10000>;
 			clocks = <&ccu CLK_BUS_HDMI1>, <&ccu CLK_HDMI_SLOW>,
-				 <&ccu 7>, <&ccu 16>;
+				 <&ccu CLK_PLL_VIDEO0>, <&ccu CLK_PLL_VIDEO1>;
 			clock-names = "bus", "mod", "pll-0", "pll-1";
 			resets = <&ccu RST_BUS_HDMI0>;
 			reset-names = "phy";
--- a/drivers/gpu/drm/sun4i/sun4i_drv.c
+++ b/drivers/gpu/drm/sun4i/sun4i_drv.c
@@ -112,6 +112,7 @@
 	}
 
 	drm_mode_config_init(drm);
+	// drm->mode_config.allow_fb_modifiers = true;
 
 	ret = component_bind_all(drm->dev, drm);
 	if (ret) {
--- a/drivers/gpu/drm/drm_probe_helper.c
+++ b/drivers/gpu/drm/drm_probe_helper.c
@@ -404,6 +404,8 @@
 
 	DRM_DEBUG_KMS("[CONNECTOR:%d:%s]\n", connector->base.id,
 			connector->name);
+	// using console args parsed to kernel to force connector on
+	// connector->force = DRM_FORCE_ON;
 
 retry:
 	ret = drm_modeset_lock(&dev->mode_config.connection_mutex, &ctx);
--- a/drivers/gpu/drm/sun4i/sun4i_rgb.c
+++ b/drivers/gpu/drm/sun4i/sun4i_rgb.c
@@ -220,7 +220,7 @@
 	ret = drm_encoder_init(drm,
 			       &rgb->encoder,
 			       &sun4i_rgb_enc_funcs,
-			       DRM_MODE_ENCODER_NONE,
+			       DRM_MODE_ENCODER_LVDS,
 			       NULL);
 	if (ret) {
 		dev_err(drm->dev, "Couldn't initialise the rgb encoder\n");
@@ -235,7 +235,7 @@
 					 &sun4i_rgb_con_helper_funcs);
 		ret = drm_connector_init(drm, &rgb->connector,
 					 &sun4i_rgb_con_funcs,
-					 DRM_MODE_CONNECTOR_Unknown);
+					 DRM_MODE_CONNECTOR_LVDS);
 		if (ret) {
 			dev_err(drm->dev, "Couldn't initialise the rgb connector\n");
 			goto err_cleanup_connector;
