--- b/arch/arm/boot/dts/sun8i-r40-bananapi-m2-ultra.dts
+++ b/arch/arm/boot/dts/sun8i-r40-bananapi-m2-ultra.dts
@@ -53,6 +53,9 @@
 	aliases {
 		ethernet0 = &gmac;
 		serial0 = &uart0;
+		mmc0 = &mmc0;
+		mmc1 = &mmc2;
+		mmc2 = &mmc1;
 	};
 
 	chosen {
@@ -106,7 +109,7 @@
 		regulator-name = "vcc5v0";
 		regulator-min-microvolt = <5000000>;
 		regulator-max-microvolt = <5000000>;
-		gpio = <&pio 7 23 GPIO_ACTIVE_HIGH>; /* PH23 */
+		gpio = <&pio 8 13 GPIO_ACTIVE_HIGH>; /* PI13 */
 		enable-active-high;
 	};
 
@@ -206,7 +209,7 @@
 	vqmmc-supply = <&reg_dcdc1>;
 	bus-width = <8>;
 	non-removable;
-	status = "okay";
+	status = "disabled";
 };
 
 &ohci1 {
@@ -334,6 +347,9 @@
 };
 
 &usbphy {
+	usb0_id_det-gpios = <&pio 8 4 GPIO_ACTIVE_HIGH>; /* PI4 */
+	usb0_vbus_det-gpios = <&pio 8 8 GPIO_ACTIVE_HIGH>; /* PI8 */
+	usb0_vbus-supply = <&reg_vcc5v0>;
 	usb1_vbus-supply = <&reg_vcc5v0>;
 	usb2_vbus-supply = <&reg_vcc5v0>;
 	status = "okay";
