--- b/arch/arm/boot/dts/sun8i-r40-bananapi-m2-ultra.dts
+++ b/arch/arm/boot/dts/sun8i-r40-bananapi-m2-ultra.dts
@@ -104,6 +104,15 @@
 		};
 	};
 
+	reg_otg_vcc5v0: vcc5v0 {
+		compatible = "regulator-fixed";
+		regulator-name = "otg_vcc5v0";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		gpio = <&pio 8 13 GPIO_ACTIVE_HIGH>; /* PI13 */
+		enable-active-high;
+	};
+
 	reg_vcc5v0: vcc5v0 {
 		compatible = "regulator-fixed";
 		regulator-name = "vcc5v0";
@@ -131,6 +140,22 @@
 	status = "okay";
 };
 
+&otg_sram {
+	status = "okay";
+};
+
+&usb_otg {
+	status = "okay";
+};
+
+&ehci0 {
+	status = "okay";
+};
+
+&ohci0 {
+	status = "okay";
+};
+
 &ehci1 {
 	status = "okay";
 };
@@ -349,7 +374,7 @@
 &usbphy {
 	usb0_id_det-gpios = <&pio 8 4 GPIO_ACTIVE_HIGH>; /* PI4 */
 	usb0_vbus_det-gpios = <&pio 8 8 GPIO_ACTIVE_HIGH>; /* PI8 */
-	usb0_vbus-supply = <&reg_vcc5v0>;
+	usb0_vbus-supply = <&reg_otg_vcc5v0>;
 	usb1_vbus-supply = <&reg_vcc5v0>;
 	usb2_vbus-supply = <&reg_vcc5v0>;
 	status = "okay";
