--- a/arch/arm/boot/dts/sun8i-r40-bananapi-m2-ultra.dts
+++ b/arch/arm/boot/dts/sun8i-r40-bananapi-m2-ultra.dts
@@ -94,11 +94,13 @@
 		user-led-green {
 			label = "bananapi:green:user";
 			gpios = <&pio 7 21 GPIO_ACTIVE_HIGH>;
+			linux,default-trigger = "heartbeat";
 		};
 
 		user-led-blue {
 			label = "bananapi:blue:user";
 			gpios = <&pio 7 22 GPIO_ACTIVE_HIGH>;
+			default-state = "on";
 		};
 	};
 
