--- a/drivers/net/sun8i_emac.c
+++ b/drivers/net/sun8i_emac.c
@@ -131,6 +131,7 @@ struct emac_variant {
 	uint			syscon_offset;
 	bool			soc_has_internal_phy;
 	bool			support_rmii;
+	bool			h3_emac_hack;
 };
 
 struct emac_dma_desc {
@@ -719,6 +720,30 @@ static int sun8i_emac_eth_probe(struct u
 	sun8i_mdio_init(dev->name, dev);
 	priv->bus = miiphy_get_dev_by_name(dev->name);
 
+	if (priv->variant->h3_emac_hack && priv->use_internal_phy)
+	{
+		printf("%s: H3 phy hacks begin\n", __func__);
+		sun8i_mdio_write(priv->bus, priv->phyaddr, MDIO_DEVAD_NONE, 0x1f, 0x013d);
+		sun8i_mdio_write(priv->bus, priv->phyaddr, MDIO_DEVAD_NONE, 0x10, 0x3ffe);
+		sun8i_mdio_write(priv->bus, priv->phyaddr, MDIO_DEVAD_NONE, 0x1f, 0x063d);
+		sun8i_mdio_write(priv->bus, priv->phyaddr, MDIO_DEVAD_NONE, 0x13, 0x8000);
+		sun8i_mdio_write(priv->bus, priv->phyaddr, MDIO_DEVAD_NONE, 0x1f, 0x023d);
+		sun8i_mdio_write(priv->bus, priv->phyaddr, MDIO_DEVAD_NONE, 0x18, 0x1000);
+		sun8i_mdio_write(priv->bus, priv->phyaddr, MDIO_DEVAD_NONE, 0x1f, 0x063d);
+		sun8i_mdio_write(priv->bus, priv->phyaddr, MDIO_DEVAD_NONE, 0x15, 0x132c);
+		sun8i_mdio_write(priv->bus, priv->phyaddr, MDIO_DEVAD_NONE, 0x1f, 0x013d);
+		sun8i_mdio_write(priv->bus, priv->phyaddr, MDIO_DEVAD_NONE, 0x13, 0xd602);
+		sun8i_mdio_write(priv->bus, priv->phyaddr, MDIO_DEVAD_NONE, 0x17, 0x003b);
+		sun8i_mdio_write(priv->bus, priv->phyaddr, MDIO_DEVAD_NONE, 0x1f, 0x063d);
+		sun8i_mdio_write(priv->bus, priv->phyaddr, MDIO_DEVAD_NONE, 0x14, 0x7088);
+		sun8i_mdio_write(priv->bus, priv->phyaddr, MDIO_DEVAD_NONE, 0x1f, 0x033d);
+		sun8i_mdio_write(priv->bus, priv->phyaddr, MDIO_DEVAD_NONE, 0x11, 0x8530);
+		sun8i_mdio_write(priv->bus, priv->phyaddr, MDIO_DEVAD_NONE, 0x1f, 0x003d);
+		//phy_val = sun8i_mdio_read(priv->bus, priv->phyaddr, MDIO_DEVAD_NONE, 0x3c);
+		//sun8i_mdio_write(priv->bus, priv->phyaddr, MDIO_DEVAD_NONE, 0x3c, phy_val & 0x02);
+		printf("%s: H3 phy hacks end\n", __func__);
+	}
+
 	return sun8i_phy_init(priv, dev);
 }
 
@@ -884,6 +909,7 @@ static const struct emac_variant emac_va
 	.syscon_offset		= 0x30,
 	.soc_has_internal_phy	= true,
 	.support_rmii		= true,
+	.h3_emac_hack		= true,
 };
 
 static const struct emac_variant emac_variant_r40 = {
