--- a/board/sunxi/board.c
+++ b/board/sunxi/board.c
@@ -767,7 +767,7 @@
 	if (!get_unique_sid(sid))
 		return;

-	for (i = 0; i < 4; i++) {
+	for (i = 0; i < 8; i++) {
 		sprintf(ethaddr, "ethernet%d", i);
 		if (!fdt_get_alias(fdt, ethaddr))
 			continue;
@@ -781,12 +781,15 @@
 			continue;

 		/* Non OUI / registered MAC address */
-		mac_addr[0] = (i << 4) | 0x02;
-		mac_addr[1] = (sid[0] >>  0) & 0xff;
+		// mac_addr[0] = (i << 4) | 0x02;
+		// mac_addr[1] = (sid[0] >>  0) & 0xff;
+		mac_addr[0] = 0xB0;
+		mac_addr[1] = 0xC9;
 		mac_addr[2] = (sid[3] >> 24) & 0xff;
 		mac_addr[3] = (sid[3] >> 16) & 0xff;
 		mac_addr[4] = (sid[3] >>  8) & 0xff;
-		mac_addr[5] = (sid[3] >>  0) & 0xff;
+		// mac_addr[5] = (sid[3] >>  0) & 0xff;
+		mac_addr[5] = ((sid[3] >>  0) + i) & 0xff;

 		eth_env_set_enetaddr(ethaddr, mac_addr);
 	}
