--- a/arch/arm/include/asm/arch-sunxi/clock_sun6i.h
+++ b/arch/arm/include/asm/arch-sunxi/clock_sun6i.h
@@ -450,7 +450,11 @@
 #if defined(CONFIG_MACH_SUN50I)
 #define MBUS_CLK_DEFAULT		0x81000002 /* PLL6x2 / 3 */
 #elif defined(CONFIG_MACH_SUN8I)
+#if defined(CONFIG_SUNXI_DRAM_DW_R40_DUAL_RANK)
+#define MBUS_CLK_DEFAULT		0x81000002 /* PLL6x2 / 3 R40 Dual Rank */
+#else
 #define MBUS_CLK_DEFAULT		0x81000003 /* PLL6 / 4 */
+#endif
 #else
 #define MBUS_CLK_DEFAULT		0x81000001 /* PLL6 / 2 */
 #endif
--- a/arch/arm/mach-sunxi/Kconfig
+++ b/arch/arm/mach-sunxi/Kconfig
@@ -251,6 +251,16 @@
 	---help---
 	Select this for sunxi SoCs with DesignWare DRAM controller with
 	32-bit memory buswidth.
+
+choice
+	prompt "DRAM Dual rank option"
+	optional
+
+config SUNXI_DRAM_DW_R40_DUAL_RANK
+	bool "R40 Dual Rank"
+
+endchoice
+
 endif

 config MACH_SUNXI_H3_H5
