--- a/arch/arm/mach-sunxi/Kconfig
+++ b/arch/arm/mach-sunxi/Kconfig
@@ -786,6 +786,12 @@
 	UART3 may used pins PE8 and PE9 on the Allwinner T113 board.
 	Enable this if you are using these pins for UART3.
 
+config UART0_PORT_G
+	bool "UART0 on PG17 and PG18 pins"
+	---help---
+	UART0 may used pins PG17 and PG18 on the Allwinner T113 board.
+	Enable this if you are using these pins for UART0.
+
 config OLD_SUNXI_KERNEL_COMPAT
 	bool "Enable workarounds for booting old kernels"
 	---help---
--- a/arch/arm/mach-sunxi/board.c
+++ b/arch/arm/mach-sunxi/board.c
@@ -147,9 +147,15 @@
 	sunxi_gpio_set_cfgpin(SUNXI_GPH(13), SUN9I_GPH_UART0);
 	sunxi_gpio_set_pull(SUNXI_GPH(13), SUNXI_GPIO_PULL_UP);
 #elif CONFIG_CONS_INDEX == 1 && defined(CONFIG_MACH_SUN8I_R528)
+#if IS_ENABLED(CONFIG_UART0_PORT_G)
+	sunxi_gpio_set_cfgpin(SUNXI_GPG(17), 7);
+	sunxi_gpio_set_cfgpin(SUNXI_GPG(18), 7);
+	sunxi_gpio_set_pull(SUNXI_GPG(18), SUNXI_GPIO_PULL_UP);
+#else
 	sunxi_gpio_set_cfgpin(SUNXI_GPE(2), 6);
 	sunxi_gpio_set_cfgpin(SUNXI_GPE(3), 6);
 	sunxi_gpio_set_pull(SUNXI_GPE(3), SUNXI_GPIO_PULL_UP);
+#endif
 #elif CONFIG_CONS_INDEX == 2 && defined(CONFIG_MACH_SUNIV)
 	sunxi_gpio_set_cfgpin(SUNXI_GPA(2), SUNIV_GPE_UART0);
 	sunxi_gpio_set_cfgpin(SUNXI_GPA(3), SUNIV_GPE_UART0);
--- a/drivers/pinctrl/sunxi/pinctrl-sunxi.c
+++ b/drivers/pinctrl/sunxi/pinctrl-sunxi.c
@@ -609,6 +609,8 @@
 	{ "spi0",	2 },	/* PC2-PC7 */
 #if IS_ENABLED(CONFIG_UART0_PORT_F)
 	{ "uart0",	3 },	/* PF2,PF4 */
+#elif IS_ENABLED(CONFIG_UART0_PORT_G)
+	{ "uart0",	7 },	/* PG17,PG18 */
 #else
 	{ "uart0",	6 },	/* PB0-PB1, PB8-PB9, PE2-PE3 */
 #endif
