--- a/include/configs/sunxi-common.h
+++ b/include/configs/sunxi-common.h
@@ -243,6 +243,12 @@
 #define BOOT_TARGET_DEVICES_DHCP(func)
 #endif
 
+#ifdef CONFIG_CMD_UBIFS
+#define BOOT_TARGET_DEVICES_UBIFS(func)	func(UBIFS, ubifs, 0, ubi, kernel)
+#else
+#define BOOT_TARGET_DEVICES_UBIFS(func)
+#endif
+
 /* FEL boot support, auto-execute boot.scr if a script address was provided */
 #define BOOTENV_DEV_FEL(devtypeu, devtypel, instance) \
 	"bootcmd_fel=" \
@@ -256,6 +262,7 @@
 #define BOOT_TARGET_DEVICES(func) \
 	func(FEL, fel, na) \
 	BOOT_TARGET_DEVICES_MMC(func) \
+	BOOT_TARGET_DEVICES_UBIFS(func) \
 	BOOT_TARGET_DEVICES_SCSI(func) \
 	BOOT_TARGET_DEVICES_USB(func) \
 	BOOT_TARGET_DEVICES_PXE(func) \
