diff -uNra a/include/configs/sunxi-common.h b/include/configs/sunxi-common.h
--- a/include/configs/sunxi-common.h	2021-12-06 18:39:35.038127434 +0800
+++ b/include/configs/sunxi-common.h	2021-12-06 19:15:32.413875179 +0800
@@ -60,6 +60,9 @@

 #define PHYS_SDRAM_0			CONFIG_SYS_SDRAM_BASE
 #define PHYS_SDRAM_0_SIZE		0x80000000 /* 2 GiB */
+#ifdef CONFIG_MACH_SUN8I_X3
+#define PHYS_SDRAM_0_SIZE		0x8000000 /* 128 MB */
+#endif

 #ifdef CONFIG_AHCI
 #define CONFIG_SYS_64BIT_LBA
diff -uNra a/arch/arm/include/asm/arch-sunxi/dram.h b/arch/arm/include/asm/arch-sunxi/dram.h
--- a/arch/arm/include/asm/arch-sunxi/dram.h	2021-12-08 11:00:23.709072012 +0800
+++ b/arch/arm/include/asm/arch-sunxi/dram.h	2021-12-08 11:10:12.707333223 +0800
@@ -21,6 +21,8 @@
 #include <asm/arch/dram_sun8i_a23.h>
 #elif defined(CONFIG_MACH_SUN8I_A33)
 #include <asm/arch/dram_sun8i_a33.h>
+#elif defined(CONFIG_MACH_SUN8I_X3)
+#include <asm/arch/dram_sun8i_a33.h>
 #elif defined(CONFIG_MACH_SUN8I_A83T)
 #include <asm/arch/dram_sun8i_a83t.h>
 #elif defined(CONFIG_SUNXI_DRAM_DW)
diff -uNra a/include/sunxi_gpio.h b/include/sunxi_gpio.h
--- a/include/sunxi_gpio.h	2021-12-08 11:00:23.709072012 +0800
+++ b/include/sunxi_gpio.h	2021-12-08 11:09:47.754677395 +0800
@@ -117,6 +117,7 @@
 #define SUN5I_GPB_UART0		2
 #define SUN8I_GPB_UART2		2
 #define SUN8I_A33_GPB_UART0	3
+#define SUN8I_X3_GPB_UART0	3
 #define SUN8I_A83T_GPB_UART0	2
 #define SUN8I_V3S_GPB_UART0	3
 #define SUN50I_GPB_UART0	4
diff -uNra a/arch/arm/include/asm/arch-sunxi/pwm.h b/arch/arm/include/asm/arch-sunxi/pwm.h
--- a/arch/arm/include/asm/arch-sunxi/pwm.h	2021-12-08 11:00:23.709072012 +0800
+++ b/arch/arm/include/asm/arch-sunxi/pwm.h	2021-12-08 11:11:11.668781397 +0800
@@ -36,7 +36,7 @@
 #define SUNXI_PWM_MUX			SUN6I_GPH_PWM
 #endif
 
-#if defined CONFIG_MACH_SUN8I_A23 || defined CONFIG_MACH_SUN8I_A33
+#if defined CONFIG_MACH_SUN8I_A23 || defined CONFIG_MACH_SUN8I_A33 || defined CONFIG_MACH_SUN8I_X3
 #define SUNXI_PWM_PIN0			SUNXI_GPH(0)
 #define SUNXI_PWM_MUX			SUN8I_GPH_PWM
 #endif
diff -uNra a/arch/arm/mach-sunxi/board.c b/arch/arm/mach-sunxi/board.c
--- a/arch/arm/mach-sunxi/board.c	2021-12-08 11:00:23.745074609 +0800
+++ b/arch/arm/mach-sunxi/board.c	2021-12-08 11:08:21.748198106 +0800
@@ -118,6 +118,10 @@
 	sunxi_gpio_set_cfgpin(SUNXI_GPB(0), SUN8I_A33_GPB_UART0);
 	sunxi_gpio_set_cfgpin(SUNXI_GPB(1), SUN8I_A33_GPB_UART0);
 	sunxi_gpio_set_pull(SUNXI_GPB(1), SUNXI_GPIO_PULL_UP);
+#elif CONFIG_CONS_INDEX == 1 && defined(CONFIG_MACH_SUN8I_X3)
+	sunxi_gpio_set_cfgpin(SUNXI_GPB(0), SUN8I_X3_GPB_UART0);
+	sunxi_gpio_set_cfgpin(SUNXI_GPB(1), SUN8I_X3_GPB_UART0);
+	sunxi_gpio_set_pull(SUNXI_GPB(1), SUNXI_GPIO_PULL_UP);
 #elif CONFIG_CONS_INDEX == 1 && defined(CONFIG_MACH_SUNXI_H3_H5)
 	sunxi_gpio_set_cfgpin(SUNXI_GPA(4), SUN8I_H3_GPA_UART0);
 	sunxi_gpio_set_cfgpin(SUNXI_GPA(5), SUN8I_H3_GPA_UART0);
diff -uNra a/arch/arm/mach-sunxi/clock_sun6i.c b/arch/arm/mach-sunxi/clock_sun6i.c
--- a/arch/arm/mach-sunxi/clock_sun6i.c	2021-12-08 11:00:23.745074609 +0800
+++ b/arch/arm/mach-sunxi/clock_sun6i.c	2021-12-08 11:03:15.559711933 +0800
@@ -318,6 +318,7 @@
 #endif
 
 #if defined(CONFIG_MACH_SUN8I_A33) || \
+    defined(CONFIG_MACH_SUN8I_X3) || \
     defined(CONFIG_MACH_SUN8I_R40) || \
     defined(CONFIG_MACH_SUN50I)
 void clock_set_pll11(unsigned int clk, bool sigma_delta_enable)
diff -uNra a/arch/arm/mach-sunxi/cpu_info.c b/arch/arm/mach-sunxi/cpu_info.c
--- a/arch/arm/mach-sunxi/cpu_info.c	2021-12-08 11:00:23.745074609 +0800
+++ b/arch/arm/mach-sunxi/cpu_info.c	2021-12-08 11:03:35.524747003 +0800
@@ -84,6 +84,8 @@
 	printf("CPU:   Allwinner A23 (SUN8I %04x)\n", sunxi_get_sram_id());
 #elif defined CONFIG_MACH_SUN8I_A33
 	printf("CPU:   Allwinner A33 (SUN8I %04x)\n", sunxi_get_sram_id());
+#elif defined CONFIG_MACH_SUN8I_X3
+	printf("CPU:   Allwinner X3 (SUN8I %04x)\n", sunxi_get_sram_id());
 #elif defined CONFIG_MACH_SUN8I_A83T
 	printf("CPU:   Allwinner A83T (SUN8I %04x)\n", sunxi_get_sram_id());
 #elif defined CONFIG_MACH_SUN8I_H3
diff -uNra a/arch/arm/cpu/armv7/sunxi/sram.c b/arch/arm/cpu/armv7/sunxi/sram.c
--- a/arch/arm/cpu/armv7/sunxi/sram.c
+++ b/arch/arm/cpu/armv7/sunxi/sram.c
@@ -32,7 +32,7 @@
 				setbits_le32(SUNXI_SRAMC_BASE + 0x44, 0x1800);
 			else /* 0x1661 ? */
 				setbits_le32(SUNXI_SRAMC_BASE + 0x44, 0xc0);
-		} else if (IS_ENABLED(CONFIG_MACH_SUN8I_A33)) {
+		} else if (IS_ENABLED(CONFIG_MACH_SUN8I_A33) || IS_ENABLED(CONFIG_MACH_SUN8I_X3)) {
 			if (version != 0x1667)
 				setbits_le32(SUNXI_SRAMC_BASE + 0x44, 0xc0);
 		}
diff -uNra a/arch/arm/mach-sunxi/Kconfig b/arch/arm/mach-sunxi/Kconfig
--- a/arch/arm/mach-sunxi/Kconfig	2021-12-08 11:00:23.745074609 +0800
+++ b/arch/arm/mach-sunxi/Kconfig	2021-12-08 11:05:40.938471848 +0800
@@ -25,7 +25,7 @@ config DRAM_SUN8I_A33
 	bool
 	help
 	  Select this dram controller driver for Sun8i platforms,
-	  for A33 SOC.
+	  for A33/X3 SOC.

 config DRAM_SUN8I_A83T
 	bool
@@ -363,6 +363,19 @@ config MACH_SUN8I_A33
 	select SYS_I2C_SUN8I_RSB
 	select ARMV7_BOOT_SEC_DEFAULT if OLD_SUNXI_KERNEL_COMPAT

+config MACH_SUN8I_X3
+	bool "sun8i (Allwinner X3)"
+	select CPU_V7A
+	select CPU_V7_HAS_NONSEC
+	select CPU_V7_HAS_VIRT
+	select ARCH_SUPPORT_PSCI
+	select DRAM_SUN8I_A33
+	select PHY_SUN4I_USB
+	select SUNXI_GEN_SUN6I
+	select SUPPORT_SPL
+	select ARMV7_BOOT_SEC_DEFAULT if OLD_SUNXI_KERNEL_COMPAT
+	imply CONS_INDEX_5 if !DM_SERIAL
+
 config MACH_SUN8I_A83T
 	bool "sun8i (Allwinner A83T)"
 	select CPU_V7A
@@ -468,13 +481,14 @@ config MACH_SUN50I_H616

 endchoice

-# The sun8i SoCs share a lot, this helps to avoid a lot of "if A23 || A33"
+# The sun8i SoCs share a lot, this helps to avoid a lot of "if A23 || A33 || X3"
 config MACH_SUN8I
 	bool
 	select SPL_ARMV7_SET_CORTEX_SMPEN if !ARM64
 	select SUN6I_PRCM
 	default y if MACH_SUN8I_A23
 	default y if MACH_SUN8I_A33
+	default y if MACH_SUN8I_X3
 	default y if MACH_SUN8I_A83T
 	default y if MACH_SUNXI_H3_H5
 	default y if MACH_SUN8I_R40
@@ -621,7 +635,8 @@ config DRAM_ZQ
 	int "sunxi dram zq value"
 	depends on !MACH_SUN50I_H616
 	default 123 if MACH_SUN4I || MACH_SUN5I || MACH_SUN6I || \
-		       MACH_SUN8I_A23 || MACH_SUN8I_A33 || MACH_SUN8I_A83T
+		       MACH_SUN8I_A23 || MACH_SUN8I_A33 || MACH_SUN8I_X3 || \
+			   MACH_SUN8I_A83T
 	default 127 if MACH_SUN7I
 	default 14779 if MACH_SUN8I_V3S
 	default 3881979 if MACH_SUNXI_H3_H5 || MACH_SUN8I_R40 || MACH_SUN50I_H6
diff -uNra a/board/sunxi/board.c b/board/sunxi/board.c
--- a/board/sunxi/board.c	2021-12-08 11:00:23.917087022 +0800
+++ b/board/sunxi/board.c	2021-12-08 11:12:39.046692643 +0800
@@ -741,7 +741,8 @@
 	 */
 #if !defined(CONFIG_MACH_SUN4I) && !defined(CONFIG_MACH_SUN5I) && \
     !defined(CONFIG_MACH_SUN6I) && !defined(CONFIG_MACH_SUN7I) && \
-    !defined(CONFIG_MACH_SUN8I_A23) && !defined(CONFIG_MACH_SUN8I_A33)
+    !defined(CONFIG_MACH_SUN8I_A23) && !defined(CONFIG_MACH_SUN8I_A33) && \
+	!defined(CONFIG_MACH_SUN8I_X3)
 	sid[3] = crc32(0, (unsigned char *)&sid[1], 12);
 #endif
 
diff -uNra a/drivers/clk/sunxi/Kconfig b/drivers/clk/sunxi/Kconfig
--- a/drivers/clk/sunxi/Kconfig	2021-12-08 11:00:24.017094236 +0800
+++ b/drivers/clk/sunxi/Kconfig	2021-12-08 11:13:53.064121503 +0800
@@ -47,10 +47,10 @@
 
 config CLK_SUN8I_A23
 	bool "Clock driver for Allwinner A23/A33"
-	default MACH_SUN8I_A23 || MACH_SUN8I_A33
+	default MACH_SUN8I_A23 || MACH_SUN8I_A33 || MACH_SUN8I_X3
 	help
 	  This enables common clock driver support for platforms based
-	  on Allwinner A23/A33 SoC.
+	  on Allwinner A23/A33/X3 SoC.
 
 config CLK_SUN8I_A83T
 	bool "Clock driver for Allwinner A83T"
diff -uNra a/drivers/power/Kconfig b/drivers/power/Kconfig
--- a/drivers/power/Kconfig	2021-12-08 11:00:24.061097405 +0800
+++ b/drivers/power/Kconfig	2021-12-08 11:13:19.359490719 +0800
@@ -57,7 +57,7 @@
 	default AXP209_POWER if MACH_SUN4I || MACH_SUN5I || MACH_SUN7I
 	default AXP221_POWER if MACH_SUN6I || MACH_SUN8I_A23 || MACH_SUN8I_A33 || MACH_SUN8I_R40
 	default AXP818_POWER if MACH_SUN8I_A83T
-	default SUNXI_NO_PMIC if MACH_SUNXI_H3_H5 || MACH_SUN50I || MACH_SUN8I_V3S
+	default SUNXI_NO_PMIC if MACH_SUNXI_H3_H5 || MACH_SUN50I || MACH_SUN8I_V3S || MACH_SUN8I_X3
 
 config SUNXI_NO_PMIC
 	bool "board without a pmic"
diff -uNra a/drivers/video/sunxi/sunxi_display.c b/drivers/video/sunxi/sunxi_display.c
--- a/drivers/video/sunxi/sunxi_display.c	2021-12-08 11:00:24.089099422 +0800
+++ b/drivers/video/sunxi/sunxi_display.c	2021-12-08 11:14:24.872688360 +0800
@@ -846,7 +846,7 @@
 		(struct sunxi_ccm_reg *)SUNXI_CCM_BASE;
 
 	/* On sun6i the drc must be clocked even when in pass-through mode */
-#ifdef CONFIG_MACH_SUN8I_A33
+#if defined CONFIG_MACH_SUN8I_A33 || defined CONFIG_MACH_SUN8I_X3
 	setbits_le32(&ccm->ahb_reset1_cfg, 1 << AHB_RESET_OFFSET_SAT);
 #endif
 	setbits_le32(&ccm->ahb_reset1_cfg, 1 << AHB_RESET_OFFSET_DRC0);
