--- a/drivers/net/phy/Kconfig	2021-12-14 18:41:58.433268883 +0800
+++ b/drivers/net/phy/Kconfig	2021-12-14 18:43:19.747605236 +0800
@@ -577,6 +577,8 @@
 	  the Reduced Gigabit Media Independent Interface(RGMII) between
 	  Ethernet physical media devices and the Gigabit Ethernet controller.
 
+source "drivers/net/phy/motorcomm/Kconfig"
+
 endif # PHYLIB
 
 config MICREL_KS8995MA
--- a/drivers/net/phy/Makefile	2021-12-14 18:42:02.881872866 +0800
+++ b/drivers/net/phy/Makefile	2021-12-14 18:44:04.604718537 +0800
@@ -98,6 +98,7 @@
 obj-$(CONFIG_MICROCHIP_PHY)	+= microchip.o
 obj-$(CONFIG_MICROCHIP_T1_PHY)	+= microchip_t1.o
 obj-$(CONFIG_MICROSEMI_PHY)	+= mscc.o
+obj-$(CONFIG_MOTORCOMM_PHY)	+= motorcomm/
 obj-$(CONFIG_NATIONAL_PHY)	+= national.o
 obj-$(CONFIG_NXP_TJA11XX_PHY)	+= nxp-tja11xx.o
 obj-$(CONFIG_QSEMI_PHY)		+= qsemi.o
