--- /dev/null
+++ b/arch/arm/boot/dts/allwinner/sun8i-t3-tlink-dly-t3v2.dtsi
@@ -0,0 +1,538 @@
+/*
+ * Copyright (C) 2021 Dirk Chang <dirk@kooiot.com>
+ *
+ */
+
+/dts-v1/;
+#include "sun8i-r40.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	aliases {
+		mmc0 = &mmc0;
+		mmc1 = &mmc2;
+		serial0 = &uart0;
+		serial1 = &uart4;
+		serial2 = &uart5;
+		serial3 = &uart3;
+		serial4 = &uart2;
+		ethernet0 = &gmac;
+		ethernet1 = &emac;
+		ethernet2 = &usb_eth1;
+		ethernet3 = &usb_eth2;
+		led-boot = &status_led;
+		led-failsafe = &status_led;
+		led-upgrade = &status_led;
+	};
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+	/* Fixed crystal dedicated to mcp251x */
+	can_clk10m: can-clock-10m {
+		compatible = "fixed-clock";
+		#clock-cells = <0>;
+		clock-frequency = <10000000>;
+		clock-output-names = "can_clk10m";
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		status_led: status-led {
+			label = "kooiot:green:status";
+			gpios = <&pio 1 2 GPIO_ACTIVE_LOW>; /* PB2 / RUN */
+			linux,default-trigger = "heartbeat";
+		};
+
+		cloud {
+			label = "kooiot:green:cloud";
+			gpios = <&pio 8 21 GPIO_ACTIVE_LOW>; /* PI21 / CLOUD */
+			default-state = "off";
+		};
+
+		bs {
+			label = "kooiot:red:bs";
+			gpios = <&pio 3 20 GPIO_ACTIVE_LOW>; /* PD20 / SIGN_R */
+			default-state = "off";
+		};
+
+		gs {
+			label = "kooiot:green:gs";
+			gpios = <&pio 3 21 GPIO_ACTIVE_LOW>; /* PD21 / SIGN_G */
+			default-state = "off";
+		};
+	};
+
+	gpio_keys {
+		compatible = "gpio-keys-polled";
+		poll-interval = <20>;
+
+		reset {
+			label = "reset";
+			linux,code = <KEY_RESTART>;
+			gpios = <&pio 4 11 GPIO_ACTIVE_LOW>; /* PE11 / FG */
+		};
+	};
+
+	gpio_export {
+		compatible = "gpio-export";
+		#size-cells = <0>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&gpio_export_pins>;
+
+		pcie_power {
+			gpio-export,name = "pcie_power";
+			gpio-export,output=<1>;
+			gpios = <&pio 4 2 GPIO_ACTIVE_HIGH>; /* PE2 / 4G_PWR */
+		};
+
+		pcie_reset {
+			gpio-export,name = "pcie_reset";
+			gpio-export,output=<0>;
+			gpios = <&pio 4 3 GPIO_ACTIVE_HIGH>; /* PE3 / 4G_RST */
+		};
+
+		can_reset {
+			gpio-export,name = "can_reset";
+			gpio-export,output=<1>;
+			gpios = <&pio 4 1 GPIO_ACTIVE_LOW>; /* PE1 / RST_CAN */
+		};
+	};
+
+	reg_usb_vcc5v0: vcc5v0 {
+		compatible = "regulator-fixed";
+		regulator-name = "vcc5v0";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		// gpio = <&pio 8 13 GPIO_ACTIVE_HIGH>; /* PI13 */
+		// enable-active-high;
+	};
+};
+
+&emac {
+	pinctrl-names = "default";
+	pinctrl-0 = <&emac_ph_pins>;
+	phy-handle = <&emac_phy>;
+	phy-supply = <&reg_dcdc1>;
+	phy-mode = "mii";
+	reset-gpios = <&pio 7 27 GPIO_ACTIVE_LOW>; /* PH27 */
+	status = "okay";
+};
+
+&emac_mdio {
+	status = "okay";
+	phy-supply = <&reg_dcdc1>;
+	allwinner,emac-fix-phy-id = <0x20005c90>;
+
+	emac_phy: ethernet-phy@1 {
+		compatible = "ethernet-phy-ieee802.3-c22";
+		reg = <1>;
+		// reset-gpios = <&pio 7 27 GPIO_ACTIVE_HIGH>; /* PH27 */
+		// reset-assert-us = <10000>;
+		// reset-deassert-us = <1000000>;
+	};
+};
+
+&gmac {
+	pinctrl-names = "default";
+	pinctrl-0 = <&gmac_mii_pins>;
+	phy-handle = <&phy1>;
+	phy-mode = "mii";
+	phy-supply = <&reg_dcdc1>;
+	snps,reset-gpio = <&pio 0 17 GPIO_ACTIVE_LOW>; /* PA17 */
+	snps,reset-delays-us = <0 10000 1000000>;
+	status = "okay";
+};
+
+&gmac_mdio {
+	phy1: ethernet-phy@1 {
+		compatible = "ethernet-phy-ieee802.3-c22";
+		reg = <1>;
+		// reset-gpios = <&pio 0 17 GPIO_ACTIVE_LOW>; /* PA17 */
+		// reset-assert-us = <10000>;
+		// reset-deassert-us = <1000000>;
+	};
+};
+
+&i2c0 {
+	status = "okay";
+
+	axp22x: pmic@34 {
+		compatible = "x-powers,axp221";
+		reg = <0x34>;
+		interrupt-parent = <&nmi_intc>;
+		interrupts = <0 IRQ_TYPE_LEVEL_LOW>;
+		swin-supply = <&reg_dcdc1>;
+		x-powers,drive-vbus-en;
+	};
+};
+
+#include "axp22x.dtsi"
+
+/* SD Card */
+&mmc0 {
+	vmmc-supply = <&reg_dcdc1>;
+	vqmmc-supply = <&reg_dcdc1>;
+	vdmmc-supply = <&reg_dcdc1>;
+	bus-width = <4>;
+	cd-gpios = <&pio 8 15 GPIO_ACTIVE_LOW>; /* PI15 */
+	status = "okay";
+};
+
+/* eMMC */
+&mmc2 {
+	vmmc-supply = <&reg_dcdc1>;
+	vqmmc-supply = <&reg_dcdc1>;
+	bus-width = <8>;
+	non-removable;
+	status = "okay";
+};
+
+&mmc2_pins {
+	/* Increase drive strength for DDR modes */
+	drive-strength = <40>;
+};
+/*
+&otg_sram {
+	status = "okay";
+};
+
+&usb_otg {
+	status = "okay";
+	dr_mode = "host";
+};
+*/
+&ehci0 {
+	status = "okay";
+};
+
+&ehci1 {
+	status = "okay";
+	#address-cells = <1>;
+	#size-cells = <0>;
+
+	xr22800: xr22800@1 {
+		compatible = "usb4e2,800";
+		reg = <1>;
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		usb_eth1: exar@1 {
+			reg = <1>;
+			local-mac-address = [00 00 00 00 00 00];
+		};
+	};
+};
+
+&ehci2 {
+	status = "okay";
+	#address-cells = <1>;
+	#size-cells = <0>;
+
+	xr22804: xr22804@1 {
+		compatible = "usb4e2,804";
+		reg = <1>;
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		usb_eth2: exar@1 {
+			reg = <1>;
+			local-mac-address = [00 00 00 00 00 00];
+		};
+		xr_uart1: xr-uart1@2 {
+			reg = <2>;
+			#address-cells = <2>;
+			#size-cells = <0>;
+			xr-uart1-interface@0 {	/* interface 0 of configuration 1 */
+				reg = <0 1>;
+				xr_485_mode;
+				xr_tty_minor=<0>;
+			};
+		};
+		xr_uart2: xr-uart2@3 {
+			reg = <3>;
+			#address-cells = <2>;
+			#size-cells = <0>;
+			xr-uart2-interface@0 {	/* interface 0 of configuration 1 */
+				reg = <0 1>;
+				xr_485_mode;
+				xr_tty_minor=<1>;
+			};
+		};
+		xr_uart3: xr-uart3@4 {
+			reg = <4>;
+			#address-cells = <2>;
+			#size-cells = <0>;
+			xr-uart3-interface@0 {	/* interface 0 of configuration 1 */
+				reg = <0 1>;
+				xr_485_mode;
+				xr_tty_minor=<2>;
+			};
+		};
+		xr_uart4: xr-uart4@5 {
+			reg = <5>;
+			#address-cells = <2>;
+			#size-cells = <0>;
+			xr-uart4-interface@0 {	/* interface 0 of configuration 1 */
+				reg = <0 1>;
+				xr_485_mode;
+				xr_tty_minor=<3>;
+			};
+		};
+	};
+};
+
+&ohci0 {
+	status = "okay";
+};
+
+&ohci1 {
+	status = "okay";
+};
+
+&ohci2 {
+	status = "okay";
+};
+
+&pio {
+	// pinctrl-names = "default";
+	// pinctrl-0 = <&clk_out_a_pin>;
+	vcc-pa-supply = <&reg_dcdc1>;
+	vcc-pc-supply = <&reg_aldo2>;
+	vcc-pd-supply = <&reg_dcdc1>;
+	vcc-pe-supply = <&reg_eldo1>;
+	vcc-pf-supply = <&reg_dcdc1>;
+	vcc-pg-supply = <&reg_dcdc1>;
+
+	uart3_ph_pins: uart3-ph-pins {
+		pins = "PH0", "PH1";
+		function = "uart3";
+	};
+
+	uart5_dly_pins: uart5-dly-pins {
+		pins = "PI10", "PH7";
+		function = "uart5";
+	};
+
+	i2c3_pins: i2c3-pins {
+		pins = "PI0", "PI1";
+		function = "i2c3";
+	};
+
+	gpio_export_pins: gpio-export-pins {
+		pins = "PE7", "PE8";
+		function = "gpio_out";
+	};
+
+	gmac_mii_pins: gmac-mii-pins {
+		pins = "PA0", "PA1", "PA2",
+			"PA3", "PA4", "PA5", "PA6",
+			"PA7", "PA8", "PA9", "PA10",
+			"PA11", "PA12", "PA13", "PA14",
+			"PA15", "PA16";
+		function = "gmac";
+		drive-strength = <40>;
+	};
+
+	spi2_pc_pins: spi2-pc-pins {
+		pins = "PC19", "PC20", "PC21", "PC22", "PB13";
+		function = "spi2";
+	};
+
+	spi2_irq_pins: spi2-irq-pins {
+		pins = "PH12", "PH13";
+		function = "irq";
+	};
+};
+
+&reg_dc1sw {
+	regulator-always-on;
+	regulator-min-microvolt = <3300000>;
+	regulator-max-microvolt = <3300000>;
+	regulator-name = "vcc-emac-phy";
+};
+
+&reg_dcdc1 {
+	regulator-always-on;
+	regulator-min-microvolt = <3300000>;
+	regulator-max-microvolt = <3300000>;
+	regulator-name = "vcc-3v3";
+};
+
+&reg_dcdc2 {
+	regulator-always-on;
+	regulator-min-microvolt = <1000000>;
+	regulator-max-microvolt = <1300000>;
+	regulator-name = "vdd-cpu";
+};
+
+&reg_dcdc3 {
+	regulator-always-on;
+	regulator-min-microvolt = <1000000>;
+	regulator-max-microvolt = <1300000>;
+	regulator-name = "vdd-sys";
+};
+
+&reg_dcdc5 {
+	regulator-always-on;
+	regulator-min-microvolt = <1500000>;
+	regulator-max-microvolt = <1500000>;
+	regulator-name = "vcc-dram";
+};
+
+&reg_aldo1 {
+	regulator-always-on;
+	regulator-min-microvolt = <3300000>;
+	regulator-max-microvolt = <3300000>;
+	regulator-name = "vcc-tvin";
+};
+
+&reg_aldo2 {
+	regulator-always-on;
+	regulator-min-microvolt = <1800000>;
+	regulator-max-microvolt = <1800000>;
+	regulator-name = "vcc-pc";
+};
+
+&reg_aldo3 {
+	regulator-always-on;
+	regulator-min-microvolt = <3000000>;
+	regulator-max-microvolt = <3000000>;
+	regulator-name = "avcc";
+};
+
+&reg_dldo1 {
+	regulator-min-microvolt = <3300000>;
+	regulator-max-microvolt = <3300000>;
+	regulator-name = "vcc-wifi-io";
+};
+
+&reg_dldo2 {
+	regulator-min-microvolt = <3300000>;
+	regulator-max-microvolt = <3300000>;
+	regulator-name = "vcc-wifi";
+};
+
+&reg_dldo3 {
+	regulator-min-microvolt = <3300000>;
+	regulator-max-microvolt = <3300000>;
+	regulator-name = "vcc-wifi-2";
+};
+
+&reg_dldo4 {
+	regulator-always-on;
+	regulator-min-microvolt = <2500000>;
+	regulator-max-microvolt = <2500000>;
+	regulator-name = "vdd2v5-sata";
+};
+
+&reg_eldo1 {
+	regulator-always-on;
+	regulator-min-microvolt = <2800000>;
+	regulator-max-microvolt = <2800000>;
+	regulator-name = "vcc-pe";
+};
+
+&reg_eldo2 {
+	regulator-always-on;
+	regulator-min-microvolt = <3300000>;
+	regulator-max-microvolt = <3300000>;
+	regulator-name = "vcc-none";
+};
+
+&reg_eldo3 {
+	regulator-always-on;
+	regulator-min-microvolt = <1200000>;
+	regulator-max-microvolt = <1200000>;
+	regulator-name = "vdd1v2-sata";
+};
+
+&uart0 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart0_pb_pins>;
+	status = "okay";
+};
+
+&uart2 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart2_pi_pins>;
+	status = "okay";
+};
+
+&uart3 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart3_ph_pins>;
+	status = "okay";
+};
+
+&uart4 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart4_ph_pins>;
+	status = "okay";
+};
+
+&uart5 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart5_dly_pins>;
+	status = "okay";
+};
+
+/*
+&reg_drivevbus {
+	regulator-name = "usb0-vbus";
+	status = "okay";
+};
+*/
+
+&usbphy {
+	// usb0_vbus-supply = <&reg_drivevbus>, <&reg_usb_vcc5v0>; /* PI13 */
+	// usb0_vbus_det-gpios = <&pio 8 13 GPIO_ACTIVE_HIGH>; /* PI14 */
+	status = "okay";
+};
+
+&i2c4 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&i2c4_pins>;
+	status = "okay";
+	rx8025t: rx8025t@32 {
+		compatible = "epson,rx8025t";
+		reg = <0x32>;
+		status = "okay";
+	};
+};
+
+&rtc {
+	status = "disabled"; // cpu rtc disabled
+};
+
+&spi2 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&spi2_pc_pins>, <&spi2_irq_pins>;
+	status = "okay";
+
+	mcp2515_can0: mcp2515@0 {
+		compatible = "microchip,mcp2515";
+		reg = <0>;
+		clocks = <&can_clk10m>;
+		interrupt-parent = <&pio>;
+		interrupts = <7 13 IRQ_TYPE_LEVEL_LOW>; /* PH13/EINT13 */
+		vdd-supply = <&reg_dcdc1>;
+		spi-max-frequency = <10000000>;
+		status = "okay";
+	};
+
+	mcp2515_can1: mcp2515@1 {
+		compatible = "microchip,mcp2515";
+		reg = <1>;
+		clocks = <&can_clk10m>;
+		interrupt-parent = <&pio>;
+		interrupts = <7 12 IRQ_TYPE_LEVEL_LOW>; /* PH12/EINT12 */
+		vdd-supply = <&reg_dcdc1>;
+		spi-max-frequency = <10000000>;
+		status = "okay";
+	};
+};
