--- a/arch/arm/boot/dts/allwinner/Makefile
+++ b/arch/arm/boot/dts/allwinner/Makefile
@@ -341,7 +341,8 @@ dtb-$(CONFIG_MACH_SUN8I) += \
 	sun8i-t113-iot-station.dtb \
 	sun8i-t113-tlink-rp-t113.dtb \
	sun8i-t113-tlink-dly-e102.dtb \
-	sun8i-t113-tlink-dly-e102-spinand.dtb
+	sun8i-t113-tlink-dly-e102-spinand.dtb \
+	sun8i-t113-tlink-e1.dtb
 dtb-$(CONFIG_MACH_SUN9I) += \
 	sun9i-a80-optimus.dtb \
 	sun9i-a80-cubieboard4.dtb
--- /dev/null
+++ b/arch/arm/boot/dts/allwinner/sun8i-t113-tlink-e1.dts
@@ -0,0 +1,420 @@
+// SPDX-License-Identifier: (GPL-2.0+ or MIT)
+// Copyright (C) 2024 Dirk Chang <dirk@kooiot.com>
+
+/dts-v1/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/leds/common.h>
+#include <dt-bindings/pwm/pwm.h>
+#include <dt-bindings/input/input.h>
+#include <dt-bindings/input/gpio-keys.h>
+#include "sun8i-t113s.dtsi"
+
+/ {
+	model = "ThingsLink E1";
+	compatible = "kooiot,tlink-e1", "allwinner,sun8i-t113";
+
+	aliases {
+		ethernet0 = &emac;
+		ethernet1 = &usb_eth1;
+		mmc0 = &mmc0;
+		mmc1 = &mmc2;
+		mmc2 = &mmc1;
+		serial0 = &uart4;
+		serial1 = &uart1;
+		serial2 = &uart2;
+		serial3 = &uart3;
+		i2c0 = &i2c0;
+		i2c1 = &i2c1;
+		i2c2 = &i2c2;
+		i2c3 = &i2c3;
+	};
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+    memory@40000000 {
+		device_type = "memory";
+		reg = <0x40000000 0x8000000>;   /* SIP 128MB */
+    };
+    
+	reg_vcc: vcc {
+		compatible = "regulator-fixed";
+		regulator-name = "vcc";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+	};
+
+	reg_usbvbus: usbvbus {
+		compatible = "regulator-fixed";
+		regulator-name = "usbvbus";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		//gpio = <&pio 3 9 GPIO_ACTIVE_LOW>; // PD19 
+		vin-supply = <&reg_vcc>;
+	};
+
+	reg_vcc_3v3: vcc-3v3 {
+		compatible = "regulator-fixed";
+		regulator-name = "vcc-3v3";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		vin-supply = <&reg_vcc>;
+	};
+
+	reg_avcc_1v8: avcc_1v8 {
+		compatible = "regulator-fixed";
+	 	regulator-name = "avcc-1v8";
+	 	regulator-min-microvolt = <1800000>;
+	 	regulator-max-microvolt = <1800000>;
+	 	vin-supply = <&reg_vcc_3v3>;
+	};
+
+	reg_vdd_cpu: vdd-cpu {
+		compatible = "regulator-fixed";
+		regulator-name = "vdd-cpu";
+		regulator-min-microvolt = <950000>;
+		regulator-max-microvolt = <950000>;
+		vin-supply = <&reg_vcc>;
+	};
+
+	reg_vcc_wifi_3v3: regulator-wifi-3v3 {
+		compatible = "regulator-fixed";
+		regulator-name = "vcc-wifi-3v3";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>; 
+		vin-supply = <&reg_vcc>;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+	 	status_led {
+	 		linux,default-trigger = "heartbeat";
+	 		label = "blue:status";
+	 		gpios = <&pio 4 11 GPIO_ACTIVE_HIGH>; /* PE11 */
+	 	};
+
+		modem {
+			label = "kooiot:green:modem";
+			gpios = <&pio 3 17 GPIO_ACTIVE_HIGH>; /* PD17 */
+			default-state = "off";
+		};
+
+		/*
+		csq {
+			label = "kooiot:green:csq";
+			gpios = <&pio 4 1 GPIO_ACTIVE_HIGH>; // PE1  Need to fix
+			default-state = "off";
+		};
+		*/
+
+		cloud {
+			label = "kooiot:green:cloud";
+			gpios = <&pio 3 18 GPIO_ACTIVE_HIGH>; /* PD18 */
+			default-state = "off";
+		};
+	};
+
+	gpio_keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "reset";
+			linux,code = <KEY_RESTART>;
+			gpios = <&pio 4 12 GPIO_ACTIVE_LOW>; /* PE12 */
+		};
+	};
+
+	gpio_export {
+		compatible = "gpio-export";
+		#size-cells = <0>;
+
+		eth1_reset {
+			gpio-export,name = "eth1_reset";
+			gpio-export,output=<0>;
+			gpios = <&pio 3 19 GPIO_ACTIVE_HIGH>; /* PD19 */
+		};
+
+		pcie_reset {
+			gpio-export,name = "pcie_reset";
+			gpio-export,output=<0>;
+			gpios = <&pio 3 3 GPIO_ACTIVE_HIGH>; /* PD3 */
+		};
+
+		pcie_power {
+			gpio-export,name = "pcie_power";
+			gpio-export,output=<1>;
+			gpios = <&pio 3 20 GPIO_ACTIVE_HIGH>; /* PD20 */
+		};
+	};
+
+	wifi_pwrseq: wifi-pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		pinctrl-names = "default";
+		pinctrl-0 = <&wifi_rst>;
+		post-power-on-delay-ms = <50>;
+		reset-gpios = <&pio 6 12 GPIO_ACTIVE_LOW>; /* PG13 */
+
+		// clocks = <&ccu CLK_FANOUT1>;
+		// clock-names = "ext_clock";
+		// assigned-clocks = <&ccu CLK_FANOUT1>;
+		// assigned-clock-rates = <32768>;
+		// pinctrl-0 = <&clk_pg11_pin>;
+		// pinctrl-names = "default";
+	};
+};
+
+&reg_ldoa {
+	regulator-name = "vcc-wifi-1v8";
+	regulator-min-microvolt = <1800000>;
+	regulator-max-microvolt = <1800000>; 
+	ldo-in-supply = <&reg_vcc_3v3>;
+};
+
+
+&dcxo {
+	clock-frequency = <24000000>;
+};
+
+/*
+&codec {
+	status = "okay";
+};
+*/
+
+&cpu0 {
+	cpu-supply = <&reg_vdd_cpu>;
+};
+
+&cpu1 {
+	cpu-supply = <&reg_vdd_cpu>;
+};
+
+&wdt {
+	status = "okay";
+};
+
+&pio {
+	vcc-pb-supply = <&reg_vcc_3v3>;
+	vcc-pc-supply = <&reg_vcc_3v3>;
+	vcc-pd-supply = <&reg_vcc_3v3>;
+	vcc-pe-supply = <&reg_vcc_3v3>;
+	vcc-pf-supply = <&reg_vcc_3v3>;
+	vcc-pg-supply = <&reg_ldoa>;
+
+	/omit-if-no-ref/
+	uart1_pd21_pins: uart1-pd21-pins {
+		pins = "PD21", "PD22";
+		function = "uart1";
+	};
+
+    uart2_pd1_pins: uart2-pd1-pins {
+        pins = "PD1", "PD2";
+        function = "uart2";
+    };
+
+	uart3_pd10_pins: uart3-pd10-pins {
+        pins = "PD10", "PD11";
+        function = "uart3";
+    };
+
+    uart4_pd7_pins: uart4-pd7-pins {
+        pins = "PD7", "PD8";
+        function = "uart4";
+    };
+
+	/omit-if-no-ref/
+	i2c3_pb_pins: i2c3-pb-pins {
+		pins = "PB6", "PB7";
+		function = "i2c3";
+	};
+
+	emac_rst_pin: emac-rst-pin {
+		pins = "PE13";
+		function = "gpio_out";
+	};
+
+	wifi_rst: wifi-rst-pin {
+		pins = "PG13";
+		function = "gpio_out";
+	};
+};
+
+&uart1 {
+	pinctrl-0 = <&uart1_pd21_pins>;
+	pinctrl-names = "default";
+	status = "okay";
+};
+
+&uart2 {
+    pinctrl-names = "default";
+    pinctrl-0 = <&uart2_pd1_pins>; 
+    status = "okay";
+};
+
+&uart3 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart3_pd10_pins>;
+	status = "okay";
+};
+
+&uart4 {
+	pinctrl-0 = <&uart4_pd7_pins>;
+	pinctrl-names = "default";
+	status = "okay";
+};
+
+&i2c3 {
+	pinctrl-0 = <&i2c3_pb_pins>;
+	pinctrl-names = "default";
+	status = "okay";
+
+	sd2059: sd2059@32 {
+		compatible = "whwave,sd3078";
+		reg = <0x32>;
+		status = "okay";
+	};
+
+	at24c02: eeprom@50 {
+		compatible = "atmel,24c02";
+		reg = <0x50>;
+		pagesize = <16>;
+	};
+};
+
+&mmc0 {
+	bus-width = <4>;
+	disable-wp;
+	vmmc-supply = <&reg_vcc_3v3>;
+	vqmmc-supply = <&reg_vcc_3v3>;
+	pinctrl-0 = <&mmc0_pins>;
+	pinctrl-names = "default";
+	cd-gpios = <&pio 5 6 GPIO_ACTIVE_LOW>; /* PF6 */
+	status = "okay";
+};
+
+&mmc1_pins {
+	drive-strength = <40>;
+	bias-pull-up;
+};
+
+&mmc1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&mmc1_pins>;
+	vmmc-supply = <&reg_vcc_3v3>;
+	vqmmc-supply = <&reg_ldoa>;
+	// mmc-pwrseq = <&wifi_pwrseq>;
+	bus-width = <4>;
+	max-frequency = <15000000>;
+	non-removable;
+	no-sd;
+	cap-sdio-irq;
+	keep-power-in-suspend;
+	ignore-pm-notify;
+	status = "disabled";
+
+	aw869a: sdio_wifi@1 {
+        reg = <1>;
+        interrupt-parent = <&pio>;
+        interrupts = <6 11 IRQ_TYPE_LEVEL_HIGH>; /* PG11 */
+        interrupt-names = "host-wake";
+	};
+};
+
+&mmc2 {
+	bus-width = <4>;
+	non-removable;
+	vmmc-supply = <&reg_vcc_3v3>;
+	vqmmc-supply = <&reg_vcc_3v3>;
+	pinctrl-0 = <&mmc2_pins>;
+	pinctrl-names = "default";
+	status = "okay";
+};
+
+&usb_otg {
+	dr_mode = "otg";
+	status = "disabled";
+};
+
+&sid {
+	chipid: chipid@0 {
+		reg = <0x0 0x10>;
+	};
+};
+
+&ehci0 {
+	status = "okay";
+	#address-cells = <1>;
+	#size-cells = <0>;
+	usb_eth1: asix@1 {
+		reg = <1>;
+		nvmem-cells = <&chipid>;
+		nvmem-cell-names = "chipid";
+		local-mac-address = [00 00 00 00 00 00];
+	};
+};
+
+&ehci1 {
+	status = "okay";
+};
+
+
+&ohci0 {
+	status = "okay";
+};
+
+&ohci1 {
+	status = "okay";
+};
+
+
+&ths {
+	status = "okay";
+	vref-supply = <&reg_ldoa>;
+};
+
+&usbphy {
+	usb0_vbus-supply = <&reg_usbvbus>;
+	usb1_vbus-supply = <&reg_usbvbus>;
+	status = "okay";
+};
+
+&mdio {
+	status = "okay";
+	ext_rmii_phy: ethernet-phy@3 {
+		compatible = "ethernet-phy-ieee802.3-c22";
+		reg = <3>;
+		tlink,yt8512b-hack;
+		motorcomm,led0 = <0x30>;
+		motorcomm,led1 = <0x330>;
+	};
+};
+
+&emac {
+	pinctrl-0 = <&rmii_pe_pins>;
+	pinctrl-names = "default";
+	phy-handle = <&ext_rmii_phy>;
+	phy-mode = "rmii";
+	phy-supply = <&reg_vcc_3v3>;
+	snps,reset-gpio = <&pio 4 13 GPIO_ACTIVE_LOW>; /* PE13 */
+	snps,reset-delays-us = <0 50000 500000>;
+	status = "okay";
+};
+
+&can0 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&can0_pins>;
+	status = "okay";
+};
+
+&can1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&can1_pins>;
+	status = "disabled";
+};
+
+&rtc {
+	status = "disabled";
+};
