--- a/arch/arm/Kconfig.debug
+++ b/arch/arm/Kconfig.debug
@@ -1113,6 +1113,14 @@ choice
 
 	config DEBUG_SUNXI_R_UART
 		bool "Kernel low-level debugging messages via sunXi R_UART"
+		depends on MACH_SUN8I || MACH_RISCV
+		select DEBUG_UART_8250
+		help
+		  Say Y here if you want kernel low-level debugging support
+		  on Allwinner D1/T113/R528 based platforms on the UART0.
+
+	config DEBUG_SUNXI_D1_UART0
+		bool "Kernel low-level debugging messages via sun8i D1 UART0"
 		depends on MACH_SUN6I || MACH_SUN8I
 		select DEBUG_UART_8250
 		help
@@ -1549,6 +1557,7 @@ config DEBUG_UART_PHYS
 	default 0x01d0c000 if DEBUG_DAVINCI_DA8XX_UART1
 	default 0x01d0d000 if DEBUG_DAVINCI_DA8XX_UART2
 	default 0x01f02800 if DEBUG_SUNXI_R_UART
+	default 0x02500000 if DEBUG_SUNXI_D1_UART0
 	default 0x02530c00 if DEBUG_KEYSTONE_UART0
 	default 0x02531000 if DEBUG_KEYSTONE_UART1
 	default 0x03010fe0 if ARCH_RPC
@@ -1698,6 +1707,7 @@ config DEBUG_UART_VIRT
 	default 0xf1c28000 if DEBUG_SUNXI_UART0
 	default 0xf1c28400 if DEBUG_SUNXI_UART1
 	default 0xf1f02800 if DEBUG_SUNXI_R_UART
+	default 0xf2500000 if DEBUG_SUNXI_D1_UART0
 	default 0xf31004c0 if DEBUG_MESON_UARTAO
 	default 0xf4090000 if DEBUG_LPC32XX
 	default 0xf4200000 if DEBUG_GEMINI
