--- a/drivers/mfd/axp20x.c
+++ b/drivers/mfd/axp20x.c
@@ -623,6 +623,9 @@
 		.num_resources	= ARRAY_SIZE(axp22x_pek_resources),
 		.resources	= axp22x_pek_resources,
 	}, {
+		.name		= "axp221-gpio",
+		.of_compatible	= "x-powers,axp813-gpio",
+	}, {
 		.name		= "axp20x-regulator",
 	}, {
 		.name		= "axp22x-adc",
--- a/arch/arm/boot/dts/axp22x.dtsi
+++ b/arch/arm/boot/dts/axp22x.dtsi
@@ -62,6 +62,12 @@
 		#io-channel-cells = <1>;
 	};
 
+	axp_gpio: gpio {
+		compatible = "x-powers,axp813-gpio";
+		gpio-controller;
+		#gpio-cells = <2>;
+	};
+
 	battery_power_supply: battery-power-supply {
 		compatible = "x-powers,axp221-battery-power-supply";
 		status = "disabled";
