--- a/drivers/net/ethernet/allwinner/sun4i-emac.c
+++ b/drivers/net/ethernet/allwinner/sun4i-emac.c
@@ -821,6 +821,8 @@
 	const char *mac_addr;
 	const struct emac_quirks *quirks;
 
+	int alias_id;
+
 	ndev = alloc_etherdev(sizeof(struct emac_board_info));
 	if (!ndev) {
 		dev_err(&pdev->dev, "could not allocate device.\n");
@@ -903,6 +905,11 @@
 		goto out_release_sram;
 	}
 
+	alias_id = of_alias_get_id(np, "ethernet");
+	if (alias_id >= 0) {
+		dev_info(&pdev->dev, "Got alias ethernet%d node:%pOF\n", alias_id, np);
+		sprintf(ndev->name, "eth%d", alias_id);
+	}
 
 	/* Read MAC-address from DT */
 	mac_addr = of_get_mac_address(np);
