--- a/arch/arm/boot/dts/sun8i-r40-bananapi-m2-ultra.dts
+++ b/arch/arm/boot/dts/sun8i-r40-bananapi-m2-ultra.dts
@@ -99,6 +99,17 @@
 		};
 	};
 
+	gpio_export {
+		compatible = "gpio-export";
+		#size-cells = <0>;
+
+		// axp_gpio_1 {
+		//	gpio-export,name = "axpio_gpio_1";
+		//	gpio-export,output=<1>;
+		//	gpios = <&axp_gpio 1 GPIO_ACTIVE_HIGH>; /* GPIO */
+		// };
+	};
+
 	reg_vcc5v0: vcc5v0 {
 		compatible = "regulator-fixed";
 		regulator-name = "vcc5v0";
@@ -207,6 +218,14 @@
 	bus-width = <4>;
 	non-removable;
 	status = "okay";
+
+	ap6212: sdio_wifi@1 {
+		compatible = "brcm,bcm4329-fmac";
+		reg = <1>;
+		// interrupt-parent = <&axp_gpio>;
+		// interrupts = <0 IRQ_TYPE_LEVEL_LOW>; /* GPIO0 */
+		// interrupt-names = "host-wake";
+	};
 };
 
 &mmc2 {
--- a/arch/arm/boot/dts/sun8i-r40-bananapi-m2-ultra.dts
+++ b/arch/arm/boot/dts/sun8i-r40-bananapi-m2-ultra.dts
@@ -52,8 +52,27 @@
 	compatible = "sinovoip,bpi-m2-ultra", "allwinner,sun8i-r40";
 
 	aliases {
+		display0 = &hdmi;
 		ethernet0 = &gmac;
 		serial0 = &uart0;
+		mmc0 = &mmc0;
+		mmc1 = &mmc2;
+		mmc2 = &mmc1;
+	};
+
+	reserved-memory {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		ranges;
+
+		/* global autoconfigured region for contiguous allocations */
+		linux,cma {
+			compatible = "shared-dma-pool";
+			reusable;
+			size = <0x2000000>; // 32M
+			alignment = <0x2000>;
+			linux,cma-default;
+		};
 	};
 
 	backlight: backlight {
@@ -91,11 +110,13 @@
 		user-led-green {
 			label = "bananapi:green:user";
 			gpios = <&pio 7 21 GPIO_ACTIVE_HIGH>;
+			linux,default-trigger = "heartbeat";
 		};
 
 		user-led-blue {
 			label = "bananapi:blue:user";
 			gpios = <&pio 7 22 GPIO_ACTIVE_HIGH>;
+			default-state = "on";
 		};
 	};
 
@@ -110,12 +131,21 @@
 		};
 	};
 
+	reg_otg_vcc5v0: vcc5v0 {
+		compatible = "regulator-fixed";
+		regulator-name = "otg_vcc5v0";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		gpio = <&pio 8 13 GPIO_ACTIVE_HIGH>; /* PI13 */
+		enable-active-high;
+	};
+
 	reg_vcc5v0: vcc5v0 {
 		compatible = "regulator-fixed";
 		regulator-name = "vcc5v0";
 		regulator-min-microvolt = <5000000>;
 		regulator-max-microvolt = <5000000>;
-		gpio = <&pio 7 23 GPIO_ACTIVE_HIGH>; /* PH23 */
+		gpio = <&pio 7 23 GPIO_ACTIVE_HIGH>; /* PH23 not used */
 		enable-active-high;
 	};
 
@@ -165,6 +195,22 @@
 	};
 };
 
+&otg_sram {
+	status = "okay";
+};
+
+&usb_otg {
+	status = "okay";
+};
+
+&ehci0 {
+	status = "okay";
+};
+
+&ohci0 {
+	status = "okay";
+};
+
 &ehci1 {
 	status = "okay";
 };
@@ -419,6 +465,9 @@
 };
 
 &usbphy {
+	usb0_id_det-gpios = <&pio 8 4 GPIO_ACTIVE_HIGH>; /* PI4 */
+	usb0_vbus_det-gpios = <&pio 8 8 GPIO_ACTIVE_HIGH>; /* PI8 */
+	// usb0_vbus-supply = <&reg_otg_vcc5v0>;
 	usb1_vbus-supply = <&reg_vcc5v0>;
 	usb2_vbus-supply = <&reg_vcc5v0>;
 	status = "okay";
