--- a/arch/arm/boot/dts/sun8i-a23-a33.dtsi	2021-12-10 17:47:18.181082708 +0800
+++ b/arch/arm/boot/dts/sun8i-a23-a33.dtsi	2021-12-10 17:55:57.716674700 +0800
@@ -472,6 +472,22 @@
 			status = "disabled";
 		};
 
+		sid: eeprom@1c23800 {
+			compatible = "allwinner,sun7i-a20-sid";
+			reg = <0x01c23800 0x200>;
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			/* Data cells */
+			chipid: chipid@00 {
+				reg = <0x00 0x10>;
+			};
+
+			ths_calibration: calib@34 {
+				reg = <0x34 0x8>;
+			};
+		};
+
 		uart0: serial@1c28000 {
 			compatible = "snps,dw-apb-uart";
 			reg = <0x01c28000 0x400>;
--- a/arch/arm/boot/dts/sun8i-a33.dtsi	2021-12-10 17:53:33.219000381 +0800
+++ b/arch/arm/boot/dts/sun8i-a33.dtsi	2021-12-10 17:56:11.394162351 +0800
@@ -250,7 +250,10 @@
 		ths: ths@1c25000 {
 			compatible = "allwinner,sun8i-a33-ths";
 			reg = <0x01c25000 0x100>;
-			#thermal-sensor-cells = <0>;
+			interrupts = <GIC_SPI 31 IRQ_TYPE_LEVEL_HIGH>;
+			nvmem-cells = <&ths_calibration>;
+			nvmem-cell-names = "calibration";
+			#thermal-sensor-cells = <1>;
 			#io-channel-cells = <0>;
 		};
 
