--- a/arch/arm/boot/dts/Makefile
+++ b/arch/arm/boot/dts/Makefile
@@ -1092,6 +1092,7 @@
 	sun8i-a33-olinuxino.dtb \
 	sun8i-a33-q8-tablet.dtb \
 	sun8i-a33-sinlinx-sina33.dtb \
+	sun8i-x3-tlink-x3.dtb \
 	sun8i-a83t-allwinner-h8homlet-v2.dtb \
 	sun8i-a83t-bananapi-m3.dtb \
 	sun8i-a83t-cubietruck-plus.dtb \
--- a/arch/arm/boot/dts/sun8i-x3-tlink-x3.dts
+++ b/arch/arm/boot/dts/sun8i-x3-tlink-x3.dts
@@ -0,0 +1,280 @@
+/*
+ * Copyright 2021 Dirk Chang <dirk@kooiot.com>
+ */
+
+/dts-v1/;
+#include "sun8i-a33.dtsi"
+#include "sunxi-common-regulators.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	model = "ThingsLink X3";
+	compatible = "kooiot,tlink-x3", "allwinner,sun8i-a33";
+
+	aliases {
+		serial0 = &uart0;
+		serial1 = &uart1;
+		serial2 = &uart3;
+		mmc0 = &mmc0;
+		mmc1 = &mmc2;
+		ethernet0 = &usb_eth0;
+		ethernet1 = &usb_eth1;
+	};
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		status {
+			label = "kooiot:green:status";
+			gpios = <&pio 4 4 GPIO_ACTIVE_HIGH>; /* PE4 */
+			linux,default-trigger = "heartbeat";
+		};
+		bs {
+			label = "kooiot:green:bs";
+			gpios = <&pio 4 6 GPIO_ACTIVE_HIGH>; /* PE6 */
+			default-state = "off";
+		};
+		gs {
+			label = "kooiot:green:gs";
+			gpios = <&pio 4 5 GPIO_ACTIVE_HIGH>; /* PE5 */
+			default-state = "off";
+		};
+		modem {
+			label = "kooiot:green:modem";
+			gpios = <&pio 4 8 GPIO_ACTIVE_HIGH>; /* PE8 */
+			default-state = "off";
+		};
+		cloud {
+			label = "kooiot:green:cloud";
+			gpios = <&pio 4 7 GPIO_ACTIVE_HIGH>; /* PE7 */
+			default-state = "off";
+		};
+	};
+
+	gpio_keys {
+		compatible = "gpio-keys-polled";
+		poll-interval = <20>;
+
+		reset {
+			label = "reset";
+			linux,code = <KEY_RESTART>;
+			gpios = <&pio 4 0 GPIO_ACTIVE_LOW>; /* PE0 */
+		};
+	};
+
+	gpio_export {
+		compatible = "gpio-export";
+		#size-cells = <0>;
+
+		eth0_reset {
+			gpio-export,name = "eth0_reset";
+			gpio-export,output=<0>;
+			gpios = <&pio 4 1 GPIO_ACTIVE_HIGH>; /* PE1 */
+		};
+
+		eth1_reset {
+			gpio-export,name = "eth1_reset";
+			gpio-export,output=<0>;
+			gpios = <&pio 4 2 GPIO_ACTIVE_HIGH>; /* PE2 */
+		};
+
+		hub_reset {
+			gpio-export,name = "hub_reset";
+			gpio-export,output=<1>;
+			gpios = <&pio 4 3 GPIO_ACTIVE_LOW>; /* PE3 */
+		};
+
+		pcie_power {
+			gpio-export,name = "pcie_power";
+			gpio-export,output=<0>;
+			gpios = <&pio 4 10 GPIO_ACTIVE_LOW>; /* PE10 */
+		};
+
+		pcie_reset {
+			gpio-export,name = "pcie_reset";
+			gpio-export,output=<0>;
+			gpios = <&pio 4 11 GPIO_ACTIVE_HIGH>; /* PE11 */
+		};
+	};
+	reg_vcc1v1: vcc1v1 {
+		compatible = "regulator-fixed";
+		regulator-name = "vcc1v1";
+		regulator-min-microvolt = <1100000>;
+		regulator-max-microvolt = <1100000>;
+	};
+
+	reg_vcc3v0: vcc3v0 {
+		compatible = "regulator-fixed";
+		regulator-name = "vcc3v0";
+		regulator-min-microvolt = <3000000>;
+		regulator-max-microvolt = <3000000>;
+	};
+
+	reg_vcc5v0: vcc5v0 {
+		compatible = "regulator-fixed";
+		regulator-name = "vcc5v0";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+	};
+};
+
+&cpu0 {
+	cpu-supply = <&reg_vcc1v1>;
+};
+
+&cpu0_opp_table {
+	opp-1104000000 {
+		opp-hz = /bits/ 64 <1104000000>;
+		opp-microvolt = <1320000>;
+		clock-latency-ns = <244144>; /* 8 32k periods */
+	};
+
+	opp-1200000000 {
+		opp-hz = /bits/ 64 <1200000000>;
+		opp-microvolt = <1320000>;
+		clock-latency-ns = <244144>; /* 8 32k periods */
+	};
+};
+
+&mmc0 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&mmc0_pins>, <&mmc0_cd_pin_tlink>;
+	vmmc-supply = <&reg_vcc3v0>;
+	bus-width = <4>;
+	// broken-cd;
+	cd-gpios = <&r_pio 0 9 GPIO_ACTIVE_LOW>; /* PL9 */
+	status = "okay";
+};
+
+&mmc2 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&mmc2_8bit_pins>;
+	vmmc-supply = <&reg_vcc3v0>;
+	bus-width = <8>;
+	non-removable;
+	cap-mmc-hw-reset;
+	status = "okay";
+};
+
+&mmc2_8bit_pins {
+	/* Increase drive strength for DDR modes */
+	drive-strength = <40>;
+	/* eMMC is missing pull-ups */
+	// bias-pull-up;
+};
+
+&pio {
+	uart3_ph_pins: uart3-ph-pins@0 {
+		pins = "PH6", "PH7";
+		function = "uart3";
+	};
+
+	key_pins: key-pins@0 {
+		pins = "PE0";
+		function = "gpio_in";
+	};
+};
+
+&r_pio {
+	mmc0_cd_pin_tlink: mmc0_cd_pin {
+		pins = "PL9";
+		function = "gpio_in";
+		// bias-pull-up;
+	};
+};
+
+&rtc {
+	status = "disabled"; // cpu rtc disabled
+};
+
+&ohci0 {
+	status = "okay";
+};
+
+&ehci0 {
+	status = "okay";
+	#address-cells = <1>;
+	#size-cells = <0>;
+
+	hub@1 {
+		compatible = "usb5e3,608";
+		reg = <1>;
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		usb_eth1: asix@1 {
+			reg = <1>;
+			nvmem-cells = <&chipid>;
+			nvmem-cell-names = "chipid";
+			local-mac-address = [00 00 00 00 00 00];
+		};
+	};
+};
+
+&uart0 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart0_pb_pins>;
+	status = "okay";
+};
+
+&uart1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart1_pg_pins>;
+	status = "okay";
+};
+
+&uart3 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart3_ph_pins>;
+	status = "okay";
+};
+
+&usb_otg {
+	dr_mode = "host";
+	status = "okay";
+	#address-cells = <1>;
+	#size-cells = <0>;
+
+	usb_eth0: asix@1 {
+		reg = <1>;
+		nvmem-cells = <&chipid>;
+		nvmem-cell-names = "chipid";
+		local-mac-address = [00 00 00 00 00 00];
+	};
+};
+
+&usbphy {
+	status = "okay";
+	usb1_vbus-supply = <&reg_vcc5v0>; /* USB1 VBUS is always on */
+};
+
+&i2c0 {
+	status = "okay";
+
+	at24c02: eeprom@50 {
+		compatible = "atmel,24c02";
+		reg = <0x50>;
+		pagesize = <16>;
+		status = "okay";
+	};
+};
+
+&i2c1 {
+	status = "okay";
+
+	sd3077: sd3077@32 {
+		compatible = "whwave,sd3078";
+		reg = <0x32>;
+		status = "okay";
+	};
+};
+
+&wdt0 {
+	sunxi,start-on-init;
+};
