--- a/arch/arm/boot/dts/Makefile
+++ b/arch/arm/boot/dts/Makefile
@@ -1277,6 +1277,7 @@
 	sun8i-r40-bananapi-m2-ultra.dtb \
 	sun8i-r40-oka40i-c.dtb \
 	sun8i-r40-tlink-ok-a40i.dtb \
+	sun8i-r40-tlink-k1.dtb \
 	sun8i-s3-elimo-initium.dtb \
 	sun8i-s3-lichee-zero-plus.dtb \
 	sun8i-s3-pinecube.dtb \
--- /dev/null
+++ b/arch/arm/boot/dts/sun8i-r40-tlink-k1.dts
@@ -0,0 +1,649 @@
+/*
+ * Copyright (C) 2021 Dirk Chang <dirk@kooiot.com>
+ *
+ */
+
+/dts-v1/;
+#include "sun8i-r40.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/pwm/pwm.h>
+#include <dt-bindings/input/input.h>
+#include <dt-bindings/pinctrl/sun4i-a10.h>
+
+/ {
+	model = "ThingsLink K1";
+	compatible = "kooiot,tlink-k1", "allwinner,sun8i-r40";
+
+	aliases {
+		mmc0 = &mmc0;
+		mmc1 = &mmc2;
+		mmc2 = &mmc1;
+		serial0 = &uart0;
+		serial1 = &uart2;
+		serial2 = &uart3;
+		serial3 = &uart4;
+		serial4 = &uart5;
+		serial5 = &uart6;
+		serial6 = &uart7;
+		display0 = &hdmi;
+		display1 = &lcd0;
+		ethernet0 = &gmac;
+		ethernet1 = &emac;
+		led-boot = &status_led;
+		led-failsafe = &status_led;
+		led-upgrade = &status_led;
+	};
+
+	reserved-memory {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		ranges;
+
+		/* global autoconfigured region for contiguous allocations */
+		linux,cma {
+			compatible = "shared-dma-pool";
+			reusable;
+			size = <0x2000000>; // 32M
+			alignment = <0x2000>;
+			linux,cma-default;
+		};
+	};
+
+	lcd_backlight: lcd-backlight {
+		compatible = "pwm-backlight";
+		pwms = <&pwm 0 50000 PWM_POLARITY_INVERTED>;
+		brightness-levels = <1 2 4 8 16 32 64 128 255>;
+		default-brightness-level = <2>;
+		power-supply = <&reg_dc1sw>;
+		enable-gpios = <&pio 1 9 GPIO_ACTIVE_HIGH>; /* LCD_BL_EN: PB9 */
+	};
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+	connector {
+		compatible = "hdmi-connector";
+		type = "a";
+
+		port {
+			hdmi_con_in: endpoint {
+				remote-endpoint = <&hdmi_out_con>;
+			};
+		};
+	};
+
+	lcd0: display {
+		compatible = "dlc,dlc0700yzg-1", "simple-panel";
+		//enable-gpio = <&pio 2 9 GPIO_ACTIVE_HIGH>; /* LCD-PWR-EN: PB9 */
+		power-supply = <&reg_dc1sw>;
+		backlight = <&lcd_backlight>;
+		//status = "disabled";
+
+		port {
+			lcd0_con_in: endpoint {
+				remote-endpoint = <&tcon_lcd0_out_lcd0_lvds0_out>;
+			};
+		};
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		status_led: status-led {
+			label = "kooiot:green:status";
+			gpios = <&pio 2 0 GPIO_ACTIVE_HIGH>; /* PC0 */
+			linux,default-trigger = "heartbeat";
+		};
+
+		modem {
+			label = "kooiot:green:modem";
+			gpios = <&pio 2 1 GPIO_ACTIVE_HIGH>; /* PC1 */
+			default-state = "off";
+		};
+
+		cloud {
+			label = "kooiot:green:cloud";
+			gpios = <&pio 2 2 GPIO_ACTIVE_HIGH>; /* PC2 */
+			default-state = "off";
+		};
+	};
+
+	gpio_keys {
+		compatible = "gpio-keys-polled";
+		poll-interval = <20>;
+
+		reset {
+			label = "reset";
+			linux,code = <KEY_RESTART>;
+			gpios = <&pio 8 6 (GPIO_ACTIVE_LOW | GPIO_PULL_UP)>; /* PI6 */
+		};
+	};
+
+	gpio_export {
+		compatible = "gpio-export";
+		#size-cells = <0>;
+
+		pcie_power {
+			gpio-export,name = "pcie_power";
+			gpio-export,output=<1>;
+			gpios = <&pio 8 4 GPIO_ACTIVE_HIGH>; /* PI4 */
+		};
+
+		pcie_reset {
+			gpio-export,name = "pcie_reset";
+			gpio-export,output=<1>;
+			gpios = <&pio 1 6 GPIO_ACTIVE_LOW>; /* PB6 */
+		};
+
+		pcie_on_off {
+			gpio-export,name = "pcie_on_off";
+			gpio-export,output=<1>;
+			gpios = <&pio 1 7 GPIO_ACTIVE_HIGH>; /* PB7 */
+		};
+
+		pcie_disable {
+			gpio-export,name = "pcie_disable";
+			gpio-export,output=<1>;
+			gpios = <&pio 1 5 GPIO_ACTIVE_HIGH>; /* PB5 */
+		};
+
+		wk2xxx_reset_1 {
+			gpio-export,name = "wk2xxx_reset_1";
+			gpio-export,output=<1>;
+			gpios = <&pio 8 9 GPIO_ACTIVE_LOW>; /* PI9 */
+		};
+
+		ap_reset {
+			gpio-export,name = "ap_reset";
+			/* gpio-export,output=<1>; */
+			gpios = <&pio 1 11 GPIO_ACTIVE_LOW>; /* PB11 */
+		};
+
+		sata_en {
+			gpio-export,name = "sata_en"; /* SATA0_ACTn/GPIO_9 */
+			gpio-export,output=<1>;
+			gpios = <&pio 8 7 GPIO_ACTIVE_HIGH>; /* PI7 */
+		};
+
+		sata_sleep {
+			gpio-export,name = "sata_sleep"; /* DEVSLP0/DEVSLP */
+			gpio-export,output=<0>;
+			gpios = <&pio 8 8 GPIO_ACTIVE_HIGH>; /* PI8 */
+		};
+	};
+
+	reg_vcc3v3_sdc: vcc3v3-sdc {
+		compatible = "regulator-fixed";
+		regulator-name = "vcc3v3-sdc";
+		regulator-always-on;
+		regulator-boot-on;
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		vin-supply = <&reg_dcdc1>;
+		gpio = <&pio 8 14 GPIO_ACTIVE_LOW>; /* PI14 CARD0-PWR*/
+		// enable-active-high;
+	};
+
+	reg_usb_vcc5v0: usb-vcc5v0 {
+		compatible = "regulator-fixed";
+		regulator-name = "vcc5v0";
+		regulator-always-on;
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		gpio = <&pio 7 13 GPIO_ACTIVE_HIGH>; /* PH13 */
+		enable-active-high;
+	};
+
+	wifi_pwrseq: wifi_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		pinctrl-names = "default";
+		pinctrl-0 = <&wifi_rst>;
+		reset-gpios = <&pio 7 6 GPIO_ACTIVE_LOW>; /* PH6 WL_RESTN */
+		post-power-on-delay-ms = <50>;
+	};
+};
+
+&rtc {
+	status = "disabled"; // cpu rtc disabled
+};
+
+&ahci {
+	ahci-supply = <&reg_dldo4>;
+	phy-supply = <&reg_eldo3>;
+	status = "okay";
+};
+
+&de {
+	status = "okay";
+};
+
+&ths {
+	allwinner,calib-offset=<(-512)>;
+};
+
+&dphy {
+	status = "okay";
+};
+
+&otg_sram {
+	status = "disabled";
+};
+
+&usb_otg {
+	status = "disabled";
+};
+
+&ehci0 {
+	status = "okay";
+};
+
+&ohci0 {
+	status = "okay";
+};
+
+&ehci1 {
+	status = "okay";
+};
+
+&ohci1 {
+	status = "okay";
+};
+
+&ehci2 {
+	status = "okay";
+};
+
+&ohci2 {
+	status = "okay";
+};
+
+&emac {
+	pinctrl-names = "default";
+	pinctrl-0 = <&emac_ph_pins>;
+	phy-handle = <&emac_phy>;
+	phy-supply = <&reg_dcdc1>;
+	phy-mode = "mii";
+	status = "okay";
+};
+
+&emac_mdio {
+	status = "okay";
+	phy-supply = <&reg_dcdc1>;
+	// reset-gpios = <&pio 7 27 GPIO_ACTIVE_LOW>; /* PH27 */
+	// reset-assert-us = <10000>;
+	// reset-deassert-us = <1000000>;
+	allwinner,emac-fix-phy-id = <0x128>;
+
+	emac_phy: ethernet-phy@3 {
+		compatible = "ethernet-phy-ieee802.3-c22";
+		reg = <3>;
+		motorcomm,led0 = <0x331>;
+		motorcomm,led1 = <0x30>;
+	};
+};
+
+&gmac {
+	pinctrl-names = "default";
+	pinctrl-0 = <&gmac_rgmii_pins>;
+	phy-handle = <&phy1>;
+	phy-mode = "rgmii";
+	phy-supply = <&reg_dcdc1>;
+	snps,reset-gpio = <&pio 0 9 GPIO_ACTIVE_LOW>; /* PA9 */
+	snps,reset-delays-us = <0 10000 1000000>;
+	status = "okay";
+};
+
+&gmac_mdio {
+	phy1: ethernet-phy@3 {
+		compatible = "ethernet-phy-ieee802.3-c22";
+		reg = <3>;
+		// reset-gpios = <&pio 0 9 GPIO_ACTIVE_LOW>; /* PA9 */
+		// reset-assert-us = <10000>;
+		// reset-deassert-us = <1000000>;
+		// interrupt-parent = <&pio>;
+		// interrupts = <&pio 0 17 GPIO_ACTIVE_LOW>; /* PA17 */
+		motorcomm,led1 = <0x1e00>;
+		motorcomm,led2 = <0x24>;
+	};
+};
+
+&hdmi {
+	status = "okay";
+};
+
+&hdmi_out {
+	hdmi_out_con: endpoint {
+		remote-endpoint = <&hdmi_con_in>;
+	};
+};
+
+&tcon_lcd0_out_lcd0_lvds0_out {
+	remote-endpoint = <&lcd0_con_in>;
+};
+
+&spi2 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&spi2_pc_pins>, <&spi_irq_pins>;
+	status = "okay";
+	spi2_serial@00 {
+		compatible = "wkmic,wk2132spi";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
+		poll_mode = <0>;
+		type = <0>;
+		enable_dma = <0>;
+		irq_gpio=<&pio 8 15 IRQ_TYPE_LEVEL_LOW>; /* PI15 */
+	};
+};
+
+&i2c0 {
+	status = "okay";
+
+	axp22x: pmic@34 {
+		compatible = "x-powers,axp221";
+		reg = <0x34>;
+		interrupt-parent = <&nmi_intc>;
+		interrupts = <0 IRQ_TYPE_LEVEL_LOW>;
+	};
+};
+
+&i2c1 {
+	status = "okay";
+};
+
+&i2c2 {
+	status = "okay";
+
+	rx8010: rx8010@32 {
+		compatible = "epson,rx8010";
+		reg = <0x32>;
+		status = "okay";
+	};
+};
+
+&i2c3 {
+	status = "okay";
+
+	gt911: touchscreen@5d {
+		compatible = "goodix,gt911";
+		reg = <0x5d>;
+		interrupt-parent = <&pio>;
+		interrupts = <8 17 IRQ_TYPE_EDGE_FALLING>; /* EINT29 (PI17) */
+		//irq-gpios = <&pio 7 15 GPIO_ACTIVE_HIGH>; /* INT (PH15) */
+		reset-gpios = <&pio 8 16 GPIO_ACTIVE_HIGH>; /* RST (PI16) */
+		//touchscreen-swapped-x-y;
+		touchscreen-inverted-x;
+		touchscreen-inverted-y;
+		touchscreen-size-x = <800>;
+		touchscreen-size-y = <480>;
+	};
+};
+
+&i2c4 {
+    status = "okay";
+
+    at24c02: eeprom@50 {
+        compatible = "atmel,24c02";
+        reg = <0x50>;
+        pagesize = <16>;
+    };
+};
+
+#include "axp22x.dtsi"
+
+/* SD Card */
+&mmc0 {
+	vmmc-supply = <&reg_vcc3v3_sdc>;
+	bus-width = <4>;
+	cd-gpios = <&pio 7 3 GPIO_ACTIVE_LOW>; /* PH3 */
+	status = "okay";
+};
+
+/* SDIO (P26) */
+&mmc1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&mmc1_pg_pins>;
+	vmmc-supply = <&reg_dldo1>;
+	vqmmc-supply = <&reg_dldo3>;
+	mmc-pwrseq = <&wifi_pwrseq>;
+	bus-width = <4>;
+	max-frequency = <16000000>;
+	non-removable;
+	status = "okay";
+
+	xr819: sdio_wifi@1 {
+		reg = <1>;
+		compatible = "xradio,xr819";
+		interrupt-parent = <&pio>;
+		interrupts = <7 7 IRQ_TYPE_EDGE_RISING>; /* PH7 */
+		interrupt-names = "host-wake";
+		local-mac-address = [dc 44 6d c0 ff ee];
+	};
+};
+
+/* eMMC */
+&mmc2 {
+	vmmc-supply = <&reg_dcdc1>;
+	vqmmc-supply = <&reg_aldo2>;
+	bus-width = <8>;
+	non-removable;
+	status = "okay";
+};
+
+&pio {
+	pinctrl-names = "default";
+	// pinctrl-0 = <&clk_out_a_pin>; // used by UART6
+	vcc-pa-supply = <&reg_dcdc1>;
+	vcc-pc-supply = <&reg_aldo2>;
+	vcc-pd-supply = <&reg_dcdc1>;
+	vcc-pe-supply = <&reg_dcdc1>;
+	vcc-pf-supply = <&reg_dcdc1>;
+	vcc-pg-supply = <&reg_dldo1>;
+
+	spi2_pc_pins: spi2-pc-pins {
+		pins = "PC19", "PC20", "PC21", "PC22";
+		function = "spi2";
+	};
+
+	spi_irq_pins: spi-irq-pins {
+		pins = "PI15";
+		function = "irq";
+	};
+
+	/omit-if-no-ref/
+	lcd0_rgb888_pins: lcd0-rgb888-pins {
+		pins = "PD0", "PD1", "PD2", "PD3",
+			   "PD4", "PD5", "PD6", "PD7",
+			   "PD8", "PD9", "PD10", "PD11",
+			   "PD12", "PD13", "PD14", "PD15",
+			   "PD16", "PD17", "PD18", "PD19",
+			   "PD20", "PD21", "PD22", "PD23",
+			   "PD24", "PD25", "PD26", "PD27";
+		function = "lcd0";
+	};
+
+	wifi_rst: wifi-rst-pin {
+		pins = "PH6";
+		function = "gpio_out";
+	};
+
+	keys_pins: key-pins {
+		pins = "PI6";
+		function = "gpio_in";
+	};
+
+	uart3_ph_pins: uart3-ph-pins {
+		pins = "PH0", "PH1";
+		function = "uart3";
+	};
+
+	uart5_pi_pins: uart5-pi-pins {
+		pins = "PI10", "PI11";
+		function = "uart5";
+	};
+
+	uart6_pi_pins: uart6-pi-pins {
+		pins = "PI12", "PI13";
+		function = "uart6";
+	};
+};
+
+&pwm {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pwm_ch0_pin>;
+	status = "okay";
+};
+
+&reg_aldo1 {
+	regulator-always-on;
+	regulator-min-microvolt = <3300000>;
+	regulator-max-microvolt = <3300000>;
+	regulator-name = "vcc-tvin";
+};
+
+&reg_aldo2 {
+	regulator-always-on;
+	regulator-min-microvolt = <1800000>;
+	regulator-max-microvolt = <1800000>;
+	regulator-name = "vcc-pc";
+};
+
+&reg_aldo3 {
+	regulator-always-on;
+	regulator-min-microvolt = <3000000>;
+	regulator-max-microvolt = <3000000>;
+	regulator-name = "avcc";
+};
+
+&reg_dc1sw {
+	regulator-always-on;
+	regulator-min-microvolt = <3300000>;
+	regulator-max-microvolt = <3300000>;
+	regulator-name = "vcc-emac-phy";
+};
+
+&reg_dcdc1 {
+	regulator-always-on;
+	regulator-min-microvolt = <3300000>;
+	regulator-max-microvolt = <3300000>;
+	regulator-name = "vcc-3v3";
+};
+
+&reg_dcdc2 {
+	regulator-always-on;
+	regulator-min-microvolt = <1000000>;
+	regulator-max-microvolt = <1300000>;
+	regulator-name = "vdd-cpu";
+};
+
+&reg_dcdc3 {
+	regulator-always-on;
+	regulator-min-microvolt = <1000000>;
+	regulator-max-microvolt = <1300000>;
+	regulator-name = "vdd-sys";
+};
+
+&reg_dcdc5 {
+	regulator-always-on;
+	regulator-min-microvolt = <1500000>;
+	regulator-max-microvolt = <1500000>;
+	regulator-name = "vcc-dram";
+};
+
+&reg_dldo1 {
+	regulator-min-microvolt = <3300000>;
+	regulator-max-microvolt = <3300000>;
+	regulator-name = "vcc-wifi-io";
+};
+
+&reg_dldo2 {
+	regulator-always-on;
+	regulator-min-microvolt = <1800000>;
+	regulator-max-microvolt = <1800000>;
+	regulator-name = "vcc-wifi";
+};
+
+&reg_dldo3 {
+	regulator-always-on;
+	regulator-min-microvolt = <1800000>;
+	regulator-max-microvolt = <1800000>;
+	regulator-name = "vcc-wifi-2";
+};
+
+&reg_dldo4 {
+	regulator-min-microvolt = <2500000>;
+	regulator-max-microvolt = <2500000>;
+	regulator-name = "vdd2v5-sata";
+};
+
+&reg_eldo3 {
+	regulator-min-microvolt = <1200000>;
+	regulator-max-microvolt = <1200000>;
+	regulator-name = "vdd1v2-sata";
+};
+
+&de {
+	status = "okay";
+	mali_supply = <&reg_dcdc3>;
+};
+
+&tcon_tv1 {
+	status = "okay";
+};
+
+&tcon_lcd0 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&lcd0_rgb888_pins>;
+	status = "okay";
+};
+
+&uart0 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart0_pb_pins>;
+	status = "okay";
+};
+
+&uart2 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart2_pi_pins>;
+	status = "okay";
+};
+
+&uart3 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart3_ph_pins>;
+	status = "okay";
+};
+
+&uart4 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart4_ph_pins>;
+	status = "okay";
+};
+
+&uart5 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart5_pi_pins>;
+	status = "okay";
+};
+
+&uart6 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart6_pi_pins>;
+	status = "okay";
+};
+
+&uart7 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart7_pi_pins>;
+	status = "okay";
+};
+
+&usbphy {
+	usb0_id_det-gpios = <&pio 1 12 GPIO_ACTIVE_HIGH>; /* PB12 */
+	usb0_vbus-supply = <&reg_usb_vcc5v0>;
+	usb1_id_det-gpios = <&pio 1 13 GPIO_ACTIVE_HIGH>; /* PB13 */
+	status = "okay";
+};
