--- a/arch/arm/boot/dts/Makefile
+++ b/arch/arm/boot/dts/Makefile
@@ -1281,6 +1281,7 @@
 	sun8i-r40-tlink-ok-a40i.dtb \
	sun8i-r40-tlink-k1.dtb \
	sun8i-r40-tlink-k2.dtb \
+	sun8i-r40-tlink-k2x.dtb \
 	sun8i-r40-tlink-dj-a40i-e.dtb \
 	sun8i-r40-tlink-qh-x40.dtb \
 	sun8i-r40-tlink-nano-a40i.dtb \
--- /dev/null
+++ b/arch/arm/boot/dts/sun8i-r40-tlink-k2x.dts
@@ -0,0 +1,617 @@
+/*
+ * Copyright (C) 2022 Dirk Chang <dirk@kooiot.com>
+ *
+ */
+
+/dts-v1/;
+#include "sun8i-r40.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/pwm/pwm.h>
+#include <dt-bindings/input/input.h>
+#include <dt-bindings/pinctrl/sun4i-a10.h>
+
+/ {
+	model = "ThingsLink K2X";
+	compatible = "kooiot,tlink-k2x", "allwinner,sun8i-r40";
+
+	aliases {
+		ethernet0 = &gmac;
+		ethernet1 = &emac;
+		serial0 = &uart0;
+		serial1 = &uart2;
+		serial2 = &uart3;
+		serial3 = &uart4;
+		serial4 = &uart5;
+		serial5 = &uart6;
+		serial6 = &uart7;
+		mmc0 = &mmc0;
+		mmc1 = &mmc2;
+		mmc2 = &mmc1;
+	};
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+	/* Fixed crystal dedicated to mcp251x */
+	can_clk8m: can-clock-8m {
+		compatible = "fixed-clock";
+		#clock-cells = <0>;
+		clock-frequency = <8000000>;
+		clock-output-names = "can_clk8m";
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		status {
+			label = "kooiot:green:status";
+			gpios = <&pio 6 7 GPIO_ACTIVE_HIGH>; /* PG7 */
+			linux,default-trigger = "heartbeat";
+		};
+
+		wifi {
+			label = "kooiot:green:wifi";
+			gpios = <&pio 6 11 GPIO_ACTIVE_HIGH>; /* PG11 */
+			default-state = "off";
+		};
+
+		modem {
+			label = "kooiot:green:modem";
+			gpios = <&pio 6 8 GPIO_ACTIVE_HIGH>; /* PG8 */
+			default-state = "off";
+		};
+
+		csq {
+			label = "kooiot:green:csq";
+			gpios = <&pio 6 9 GPIO_ACTIVE_HIGH>; /* PG9 */
+			default-state = "off";
+		};
+
+		cloud {
+			label = "kooiot:green:cloud";
+			gpios = <&pio 6 10 GPIO_ACTIVE_HIGH>; /* PG10 */
+			default-state = "off";
+		};
+	};
+
+	gpio_keys {
+		compatible = "gpio-keys-polled";
+		poll-interval = <20>;
+
+		reset {
+			label = "reset";
+			linux,code = <KEY_RESTART>;
+			gpios = <&pio 8 6 (GPIO_ACTIVE_LOW | GPIO_PULL_UP)>; /* PI6 */
+		};
+	};
+
+	gpio_export {
+		compatible = "gpio-export";
+		#size-cells = <0>;
+
+		pcie_power {
+			gpio-export,name = "pcie_power";
+			gpio-export,output=<1>;
+			gpios = <&pio 8 4 GPIO_ACTIVE_HIGH>; /* PI4 */
+		};
+
+		pcie_reset {
+			gpio-export,name = "pcie_reset";
+			gpio-export,output=<1>;
+			gpios = <&pio 1 6 GPIO_ACTIVE_LOW>; /* PB6 */
+		};
+
+		pcie_card_power_off {
+			gpio-export,name = "pcie_card_power_off";
+			gpio-export,output=<1>;
+			gpios = <&pio 1 7 GPIO_ACTIVE_LOW>; /* PB7 */
+		};
+
+		pcie_disable {
+			gpio-export,name = "pcie_disable"; /* fly-mode */
+			gpio-export,output=<1>;
+			gpios = <&pio 1 5 GPIO_ACTIVE_LOW>; /* PB5 */
+		};
+
+		pcie_on_off {
+			gpio-export,name = "pcie_on_off"; /* DEVSLP0/DEVSLP pin38 */
+			gpio-export,output=<1>;
+			gpios = <&pio 8 8 GPIO_ACTIVE_HIGH>; /* PI8 */
+		};
+
+		sata_en {
+			gpio-export,name = "sata_en"; /* SATA0_ACTn/GPIO_9 pin10 */
+			gpio-export,output=<1>;
+			gpios = <&pio 8 7 GPIO_ACTIVE_HIGH>; /* PI7 */
+		};
+
+		ex_uart_reset {
+			gpio-export,name = "ex_uart_reset";
+			gpio-export,output=<1>;
+			gpios = <&pio 8 9 GPIO_ACTIVE_LOW>; /* PI9 */
+		};
+
+		can_reset {
+			gpio-export,name = "can_reset";
+			gpio-export,output=<1>;
+			gpios = <&pio 8 5 GPIO_ACTIVE_LOW>; /* PI5 */
+		};
+
+		ap_reset {
+			gpio-export,name = "ap_reset";
+			/* gpio-export,output=<1>; */
+			gpios = <&pio 1 11 GPIO_ACTIVE_LOW>; /* PB11 */
+		};
+	};
+
+	reg_vcc3v3_sdc: vcc3v3-sdc {
+		compatible = "regulator-fixed";
+		regulator-name = "vcc3v3-sdc";
+		regulator-always-on;
+		regulator-boot-on;
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		vin-supply = <&reg_dcdc1>;
+		gpio = <&pio 8 14 GPIO_ACTIVE_LOW>; /* PI14 CARD0-PWR*/
+		// enable-active-high;
+	};
+
+	reg_usb_vcc5v0: usb-vcc5v0 {
+		compatible = "regulator-fixed";
+		regulator-name = "vcc5v0";
+		regulator-always-on;
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		// gpio = <&pio 8 16 GPIO_ACTIVE_HIGH>; /* PI16 */
+		// enable-active-high;
+	};
+
+	wifi_pwrseq: wifi_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		pinctrl-names = "default";
+		pinctrl-0 = <&wifi_rst>;
+		reset-gpios = <&pio 7 6 GPIO_ACTIVE_LOW>; /* PH6 WL_RESTN */
+		post-power-on-delay-ms = <50>;
+	};
+
+	reg_vcc_wifi_3v3: vcc-wifi-3v3 {
+		compatible = "regulator-fixed";
+		regulator-name = "vcc-wifi-3v3";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		vin-supply = <&reg_eldo1 &reg_eldo2 &reg_eldo3>; /* VCC-WIFI */
+	};
+
+	reg_vdd_io_3v3: vdd-io-3v3 {
+		compatible = "regulator-fixed";
+		regulator-name = "vcc-wifi-3v3";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		vin-supply = <&reg_dldo1 &reg_dldo2 &reg_dldo3>;
+	};
+};
+
+&ahci {
+	ahci-supply = <&reg_dldo4>;
+	phy-supply = <&reg_dc5ldo>;
+	status = "okay";
+};
+
+&rtc {
+	status = "disabled"; // cpu rtc disabled
+};
+
+&ths {
+	status = "okay";
+	// allwinner,calib-offset=<(-512)>;
+};
+
+&ehci0 {
+	status = "okay";
+};
+
+&ohci0 {
+	status = "okay";
+};
+
+&ehci1 {
+	status = "okay";
+};
+
+&ohci1 {
+	status = "okay";
+};
+
+&ehci2 {
+	status = "okay";
+};
+
+&ohci2 {
+	status = "okay";
+};
+
+&emac {
+	pinctrl-names = "default";
+	pinctrl-0 = <&emac_ph_pins>;
+	phy-handle = <&emac_phy>;
+	phy-supply = <&reg_vdd_io_3v3>;
+	phy-mode = "mii";
+	status = "okay";
+};
+
+&emac_mdio {
+	status = "okay";
+	phy-supply = <&reg_vdd_io_3v3>;
+	allwinner,emac-fix-phy-id = <0x128>;
+
+	emac_phy: ethernet-phy@3 {
+		compatible = "ethernet-phy-ieee802.3-c22";
+		reg = <3>;
+		motorcomm,led0 = <0x331>;
+		motorcomm,led1 = <0x30>;
+
+		reset-gpios = <&pio 7 27 GPIO_ACTIVE_LOW>; /* PH27 */
+		reset-assert-us = <10000>;
+		reset-deassert-us = <1000000>;
+	};
+};
+
+&gmac {
+	pinctrl-names = "default";
+	pinctrl-0 = <&gmac_rgmii_pins>;
+	phy-handle = <&phy1>;
+	phy-mode = "rgmii";
+	phy-supply = <&reg_vdd_io_3v3>;
+	snps,reset-gpio = <&pio 0 9 GPIO_ACTIVE_LOW>; /* PA9 */
+	snps,reset-delays-us = <0 10000 1000000>;
+	status = "okay";
+};
+
+&gmac_mdio {
+	phy1: ethernet-phy@3 {
+		compatible = "ethernet-phy-ieee802.3-c22";
+		reg = <3>;
+		tlink,yt8521s-hack;
+		// reset-gpios = <&pio 0 9 GPIO_ACTIVE_LOW>; /* PA9 */
+		// reset-assert-us = <10000>;
+		// reset-deassert-us = <1000000>;
+		// interrupt-parent = <&pio>;
+		// interrupts = <&pio 0 17 GPIO_ACTIVE_LOW>; /* PA17 */
+		motorcomm,led1 = <0x1e00>;
+		motorcomm,led2 = <0x24>;
+	};
+};
+
+&spi2 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&spi2_pc_pins>, <&spi2_irq_pins>;
+	status = "okay";
+
+	spi2_serial: spi2-serial@0 {
+		compatible = "wkmic,wk2132spi";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
+		poll_mode = <0>;
+		type = <0>;
+		enable_dma = <0>;
+		irq_gpio=<&pio 8 15 IRQ_TYPE_LEVEL_LOW>; /* PI15/EINT27 */
+		status = "okay";
+	};
+
+	mcp2515: mcp2515@0 {
+		compatible = "microchip,mcp2515";
+		reg = <1>;
+		clocks = <&can_clk8m>;
+		interrupt-parent = <&pio>;
+		interrupts = <7 2 IRQ_TYPE_LEVEL_LOW>; /* PH2/EINT2 */
+		vdd-supply = <&reg_vcc_wifi_3v3>; /* VCC-WIFI */
+		spi-max-frequency = <10000000>;
+		status = "okay";
+	};
+};
+
+&i2c0 {
+	status = "okay";
+
+	axp22x: pmic@34 {
+		compatible = "x-powers,axp221";
+		reg = <0x34>;
+		interrupt-parent = <&nmi_intc>;
+		interrupts = <0 IRQ_TYPE_LEVEL_LOW>;
+	};
+};
+
+&i2c1 {
+	status = "okay";
+};
+
+&i2c2 {
+	status = "okay";
+
+	sd3077: sd3077@32 {
+		compatible = "whwave,sd3078";
+		reg = <0x32>;
+		status = "okay";
+	};
+
+	at24c02: eeprom@50 {
+		compatible = "atmel,24c02";
+		reg = <0x50>;
+		pagesize = <16>;
+	};
+};
+
+&i2c3 {
+	status = "okay";
+};
+
+&i2c4 {
+	status = "okay";
+};
+
+#include "axp22x.dtsi"
+
+&ac_power_supply {
+	status = "okay";
+};
+
+&battery_power_supply {
+	status = "okay";
+};
+
+&reg_drivevbus {
+	status = "okay";
+};
+
+/* &axp_led {
+	status = "okay";
+}; */
+
+/* SD Card */
+&mmc0 {
+	vmmc-supply = <&reg_vcc3v3_sdc>;
+	bus-width = <4>;
+	cd-gpios = <&pio 7 3 GPIO_ACTIVE_LOW>; /* PH3 */
+	status = "okay";
+};
+
+/* SDIO (P26) */
+&mmc1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&mmc1_pg_pins>;
+	vmmc-supply = <&reg_vcc_wifi_3v3>;
+	vqmmc-supply = <&reg_aldo1>;
+	mmc-pwrseq = <&wifi_pwrseq>;
+	bus-width = <4>;
+	max-frequency = <16000000>;
+	non-removable;
+	status = "okay";
+
+	xr819: sdio_wifi@1 {
+		reg = <1>;
+		compatible = "xradio,xr819";
+		interrupt-parent = <&pio>;
+		interrupts = <7 7 IRQ_TYPE_EDGE_RISING>; /* PH7 */
+		interrupt-names = "host-wake";
+		local-mac-address = [dc 44 6d c0 ff ee];
+	};
+};
+
+/* eMMC */
+&mmc2 {
+	vmmc-supply = <&reg_vcc_wifi_3v3>;
+	vqmmc-supply = <&reg_aldo2>;
+	bus-width = <8>;
+	non-removable;
+	status = "okay";
+};
+
+&pio {
+	pinctrl-names = "default";
+	// pinctrl-0 = <&clk_out_a_pin>; // used by UART6
+	vcc-pa-supply = <&reg_dcdc1>;
+	vcc-pc-supply = <&reg_dcdc1>;
+	vcc-pd-supply = <&reg_dcdc1>;
+	vcc-pe-supply = <&reg_dcdc1>;
+	vcc-pf-supply = <&reg_dcdc1>;
+	vcc-pg-supply = <&reg_vcc_wifi_3v3>;
+
+	spi2_pc_pins: spi2-pc-pins {
+		pins = "PC19", "PC20", "PC21", "PC22", "PB13";
+		function = "spi2";
+	};
+
+	spi2_irq_pins: spi2-irq-pins {
+		pins = "PI15", "PH2";
+		function = "irq";
+	};
+
+	wifi_rst: wifi-rst-pin {
+		pins = "PH6";
+		function = "gpio_out";
+	};
+
+	keys_pins: key-pins {
+		pins = "PI6";
+		function = "gpio_in";
+	};
+
+	uart3_ph_pins: uart3-ph-pins {
+		pins = "PH0", "PH1";
+		function = "uart3";
+	};
+
+	uart5_pi_pins: uart5-pi-pins {
+		pins = "PI10", "PI11";
+		function = "uart5";
+	};
+
+	uart6_pi_pins: uart6-pi-pins {
+		pins = "PI12", "PI13";
+		function = "uart6";
+	};
+};
+
+&pwm {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pwm_ch0_pin>;
+	status = "okay";
+};
+
+&reg_aldo1 {
+	regulator-always-on;
+	regulator-min-microvolt = <1800000>;
+	regulator-max-microvolt = <1800000>;
+	regulator-name = "vcc-wifi-1v8";
+};
+
+&reg_aldo2 {
+	regulator-always-on;
+	regulator-min-microvolt = <1800000>;
+	regulator-max-microvolt = <1800000>;
+	regulator-name = "vdd1v8-lpddr";
+};
+
+&reg_aldo3 {
+	regulator-always-on;
+	regulator-min-microvolt = <3000000>;
+	regulator-max-microvolt = <3000000>;
+	regulator-name = "avcc";  // vcc-pll-avcc
+};
+
+&reg_dc1sw {
+	regulator-always-on;
+	regulator-min-microvolt = <3300000>;
+	regulator-max-microvolt = <3300000>;
+	regulator-name = "vcc-emac-phy";
+};
+
+&reg_dcdc1 {
+	regulator-always-on;
+	regulator-min-microvolt = <3300000>;
+	regulator-max-microvolt = <3300000>;
+	regulator-name = "vcc-3v3";
+};
+
+&reg_dcdc2 {
+	regulator-always-on;
+	regulator-min-microvolt = <1000000>;
+	regulator-max-microvolt = <1300000>;
+	regulator-name = "vdd-sys";
+};
+
+&reg_dcdc3 {
+	regulator-always-on;
+	regulator-min-microvolt = <1000000>;
+	regulator-max-microvolt = <1300000>;
+	regulator-name = "vdd-cpu";
+};
+
+&reg_dcdc5 {
+	regulator-always-on;
+	regulator-min-microvolt = <1200000>;
+	regulator-max-microvolt = <1200000>;
+	regulator-name = "vcc-dram";
+};
+
+&reg_dc5ldo {
+	regulator-min-microvolt = <1200000>;
+	regulator-max-microvolt = <1200000>;
+	regulator-name = "vdd1v2-sata";
+};
+
+&reg_dldo1 {
+	regulator-min-microvolt = <3300000>;
+	regulator-max-microvolt = <3300000>;
+	regulator-name = "vdd3v3-io";
+};
+
+&reg_dldo2 {
+	regulator-always-on;
+	regulator-min-microvolt = <3300000>;
+	regulator-max-microvolt = <3300000>;
+	regulator-name = "vdd3v3-io-2";
+};
+
+&reg_dldo3 {
+	regulator-always-on;
+	regulator-min-microvolt = <3300000>;
+	regulator-max-microvolt = <3300000>;
+	regulator-name = "vdd3v3-io-3";
+};
+
+&reg_dldo4 {
+	regulator-min-microvolt = <2500000>;
+	regulator-max-microvolt = <2500000>;
+	regulator-name = "vdd2v5-sata";
+};
+
+&reg_eldo1 {
+	regulator-always-on;
+	regulator-min-microvolt = <3300000>;
+	regulator-max-microvolt = <3300000>;
+	regulator-name = "vcc-wifi";
+};
+
+&reg_eldo2 {
+	regulator-always-on;
+	regulator-min-microvolt = <3300000>;
+	regulator-max-microvolt = <3300000>;
+	regulator-name = "vcc-wifi-2";
+};
+
+&reg_eldo3 {
+	regulator-always-on;
+	regulator-min-microvolt = <3300000>;
+	regulator-max-microvolt = <3300000>;
+	regulator-name = "vcc-wifi-3";
+};
+
+&uart0 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart0_pb_pins>;
+	status = "okay";
+};
+
+&uart2 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart2_pi_pins>;
+	status = "okay";
+};
+
+&uart3 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart3_ph_pins>;
+	status = "okay";
+};
+
+&uart4 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart4_ph_pins>;
+	status = "okay";
+};
+
+&uart5 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart5_pi_pins>;
+	status = "okay";
+};
+
+&uart6 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart6_pi_pins>;
+	status = "okay";
+};
+
+&uart7 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart7_pi_pins>;
+	status = "okay";
+};
+
+&usbphy {
+	usb0_vbus-supply = <&reg_usb_vcc5v0>;
+	status = "okay";
+};
