--- a/arch/arm/boot/dts/Makefile
+++ b/arch/arm/boot/dts/Makefile
@@ -1283,6 +1283,7 @@
	sun8i-r40-tlink-dj-a40i-e.dtb \
	sun8i-r40-tlink-qh-x40.dtb \
	sun8i-r40-tlink-nano-a40i.dtb \
+	sun8i-s3-tlink-s1.dtb \
 	sun8i-s3-elimo-initium.dtb \
 	sun8i-s3-lichee-zero-plus.dtb \
 	sun8i-s3-pinecube.dtb \
--- /dev/null
+++ b/arch/arm/boot/dts/sun8i-s3-tlink-s1.dts
@@ -0,0 +1,222 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright (C) 2021 Dirk Chang <dirk@kooiot.com>
+ */
+
+/dts-v1/;
+#include "sun8i-v3.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	model = "KooIoT TLink S1";
+	compatible = "kooiot,tlink-s1", "sochip,s3",
+		     "allwinner,sun8i-v3";
+
+	aliases {
+		mmc0 = &mmc0;
+		mmc1 = &mmc2;
+		mmc2 = &mmc1;
+		serial0 = &uart0;
+		serial1 = &uart2;
+		ethernet0 = &emac;
+		ethernet1 = &xr819;
+		led-boot = &status_led;
+		led-failsafe = &status_led;
+		led-upgrade = &status_led;
+	};
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		status_led: status-led {
+			label = "kooiot:green:status";
+			gpios = <&pio 6 6 GPIO_ACTIVE_LOW>; /* PG6 */
+			linux,default-trigger = "heartbeat";
+		};
+		modem {
+			label = "kooiot:green:modem";
+			gpios = <&pio 6 7 GPIO_ACTIVE_LOW>; /* PG7 */
+			default-state = "off";
+		};
+	};
+
+	gpio_keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "reset";
+			linux,code = <KEY_RESTART>;
+			gpios = <&pio 6 11 GPIO_ACTIVE_LOW>; /* PG11 */
+		};
+	};
+
+	gpio_export {
+		compatible = "gpio-export";
+		#size-cells = <0>;
+
+		pcie_reset {
+			gpio-export,name = "pcie_reset";
+			gpio-export,output=<0>;
+			gpios = <&pio 6 10 GPIO_ACTIVE_HIGH>; /* PG10 */
+		};
+		pcie_power {
+			gpio-export,name = "pcie_power";
+			gpio-export,output=<0>;
+			gpios = <&pio 6 8 GPIO_ACTIVE_LOW>; /* PG8 */
+		};
+		wifi_power {
+			gpio-export,name = "wifi_power";
+			gpio-export,output=<0>;
+			gpios = <&pio 6 9 GPIO_ACTIVE_LOW>; // PG9 which controls vdd_wifi (1.8v)
+		};
+		wifi_reset {
+			gpio-export,name = "wifi_reset";
+			gpio-export,output=<0>;
+			gpios = <&pio 6 12 GPIO_ACTIVE_HIGH>; // PG12 controls vcc_wifi (3.3v)
+		};
+	};
+
+	reg_vcc3v3: reg-vcc3v3 {
+		compatible = "regulator-fixed";
+		regulator-name = "vcc3v3";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+	};
+
+	reg_vdd_wifi: reg-vdd-wifi {
+		compatible = "regulator-fixed";
+		regulator-name = "vdd-wifi-1";
+		regulator-min-microvolt = <1800000>;
+		regulator-max-microvolt = <1800000>;
+		// gpio = <&pio 6 9 GPIO_ACTIVE_LOW>; /* PG9 */
+	};
+
+	reg_vcc_wifi: reg-vcc-wifi {
+		compatible = "regulator-fixed";
+		regulator-name = "vcc-wifi-1";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		// gpio = <&pio 6 12 GPIO_ACTIVE_LOW>; /* PG12 */
+	};
+};
+
+&mmc0 {
+	vmmc-supply = <&reg_vcc3v3>;
+	bus-width = <4>;
+	cd-gpios = <&pio 5 6 GPIO_ACTIVE_LOW>; /* PF6 */
+	status = "okay";
+};
+
+&mmc1 {
+	vmmc-supply = <&reg_vcc_wifi>;
+	vqmmc-supply = <&reg_vdd_wifi>;
+	bus-width = <4>;
+	max-frequency = <16000000>;
+	non-removable;
+	status = "okay";
+
+	xr819: sdio_wifi@1 {
+		reg = <1>;
+		compatible = "xradio,xr819";
+		interrupt-parent = <&pio>;
+		interrupts = <6 13 IRQ_TYPE_EDGE_RISING>; /* PG13/EINT13 */
+		interrupt-names = "host-wake";
+		nvmem-cells = <&chipid>;
+		nvmem-cells-names = "chipid";
+		local-mac-address = [00 00 00 00 00 00];
+	};
+};
+
+&mmc2 {
+	compatible = "allwinner,sun8i-s3-emmc";
+	pinctrl-names = "default";
+	pinctrl-0 = <&mmc2_pins>;
+	vmmc-supply = <&reg_vcc3v3>;
+	vqmmc-supply = <&reg_vcc3v3>;
+	bus-width = <8>;
+	non-removable;
+	status = "okay";
+};
+
+&pio {
+	vcc-pb-supply = <&reg_vcc3v3>;
+	vcc-pc-supply = <&reg_vcc3v3>;
+	vcc-pf-supply = <&reg_vcc3v3>;
+	vcc-pg-supply = <&reg_vcc3v3>;
+
+	mmc2_pins: mmc2-pins {
+		pins = "PC0", "PC1", "PC2",
+			   "PC3", "PC4", "PC5",
+			   "PC6", "PC7", "PC8",
+			   "PC9", "PC10";
+		function = "mmc2";
+		drive-strength = <40>;
+		bias-pull-up;
+	};
+};
+
+&rtc {
+	status = "disabled"; // cpu rtc disabled
+};
+
+&uart0 {
+	pinctrl-0 = <&uart0_pb_pins>;
+	pinctrl-names = "default";
+	status = "okay";
+};
+
+&uart2 {
+	status = "okay";
+};
+
+&i2c0 {
+    status = "okay";
+
+    sd3077: sd3077@32 {
+        compatible = "whwave,sd3078";
+        reg = <0x32>;
+    };
+
+    at24c02: eeprom@50 {
+        compatible = "atmel,24c02";
+        reg = <0x50>;
+        pagesize = <16>;
+    };
+};
+
+&emac {
+	phy-handle = <&int_mii_phy>;
+	phy-mode = "mii";
+	allwinner,leds-active-low;
+	status = "okay";
+};
+
+/*
+&usb_otg {
+	dr_mode = "host";
+	status = "okay";
+};
+*/
+
+&ehci0 {
+	phys = <&usbphy 0>;
+	phy-names = "usb";
+	status = "okay";
+};
+
+&ohci0 {
+	phys = <&usbphy 0>;
+	phy-names = "usb";
+	status = "okay";
+};
+
+&usbphy {
+	usb0_vbus-supply = <&reg_vcc3v3>;
+	status = "okay";
+};
