--- a/drivers/net/ethernet/ralink/esw_rt3050.c
+++ b/drivers/net/ethernet/ralink/esw_rt3050.c
@@ -643,6 +643,7 @@
 		int i;

 		esw_reset_ephy(esw);
+		udelay(5000);

 		/* set the led polarity */
 		esw_w32(esw, esw->reg_led_polarity & 0x1F,
@@ -679,6 +680,11 @@
 		rt305x_mii_write(esw, 0, 28, 0x0233);
 		rt305x_mii_write(esw, 0, 29, 0x000a);
 		rt305x_mii_write(esw, 0, 30, 0x0000);
+
+		/* Fixes from XingChuang */
+		rt305x_mii_write(esw, 0, 31, 0x4000); /* change G4 page */
+		rt305x_mii_write(esw, 0, 29, 0x000d);
+		rt305x_mii_write(esw, 0, 30, 0x0500);
 	} else {
 		rt305x_mii_write(esw, 0, 31, 0x8000);
 		for (i = 0; i < 5; i++) {
