--- a/wsm.c
+++ b/wsm.c
@@ -15,6 +15,7 @@
 #include <linux/delay.h>
 #include <linux/sched.h>
 #include <linux/random.h>
+#include <linux/version.h>
 
 #include "xradio.h"
 #include "wsm.h"
@@ -23,6 +24,12 @@
 #include "sta.h"
 #include "rx.h"
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0)
+#define PRANDOM_U32 get_random_u32
+#else
+#define PRANDOM_U32 prandom_u32
+#endif
+
 #define WSM_CMD_TIMEOUT		(2 * HZ) /* With respect to interrupt loss */
 #define WSM_CMD_JOIN_TIMEOUT	(7 * HZ) /* Join timeout is 5 sec. in FW   */
 #define WSM_CMD_START_TIMEOUT	(7 * HZ)
@@ -2609,7 +2616,7 @@
 		edca = &priv->edca.params[i];
 		score = ((edca->aifns + edca->cwMin) << 16) +
 				(edca->cwMax - edca->cwMin) *
-				(prandom_u32() & 0xFFFF);
+				(PRANDOM_U32() & 0xFFFF);
 		if (score < best && (winner < 0 || i != 3)) {
 			best = score;
 			winner = i;
