--- a/main.c
+++ b/main.c
@@ -501,8 +501,8 @@
 	int if_id;
 	struct ieee80211_hw *dev;
 	struct xradio_common *hw_priv;
-	unsigned char randomaddr[ETH_ALEN];
-	const unsigned char *addr = NULL;
+	u8 addr[ETH_ALEN];
+	int rc = -1;
 
 	//init xradio_common
 	dev = xradio_init_common(sizeof(struct xradio_common));
@@ -517,12 +517,11 @@
 
 	// fill in mac addresses
 	if (hw_priv->pdev->of_node) {
-		addr = of_get_mac_address(hw_priv->pdev->of_node);
+		rc = of_get_mac_address(hw_priv->pdev->of_node, addr);
 	}
 	if (!addr) {
 		dev_warn(hw_priv->pdev, "no mac address provided, using random\n");
-		eth_random_addr(randomaddr);
-		addr = randomaddr;
+		eth_random_addr(addr);
 	}
 	memcpy(hw_priv->addresses[0].addr, addr, ETH_ALEN);
 	memcpy(hw_priv->addresses[1].addr, addr, ETH_ALEN);
