--- a/src/r8168_n.c
+++ b/src/r8168_n.c
@@ -28043,6 +28043,8 @@ rtl8168_init_board(struct pci_dev *pdev,
 
         SET_MODULE_OWNER(dev);
         SET_NETDEV_DEV(dev, &pdev->dev);
+		// https://patchew.org/linux/20250717090037.4097520-1-primoz.fiser@norik.com/20250717090037.4097520-2-primoz.fiser@norik.com/
+		dev->dev.of_node = pdev->dev.of_node;
         tp = netdev_priv(dev);
         tp->dev = dev;
         tp->pci_dev = pdev;
