--- a/drivers/usb/musb/sunxi.c	2021-12-11 12:03:48.526656696 +0800
+++ b/drivers/usb/musb/sunxi.c	2021-12-11 11:17:47.704256199 +0800
@@ -789,6 +789,7 @@
 	pinfo.num_res	= pdev->num_resources;
 	pinfo.data	= &pdata;
 	pinfo.size_data = sizeof(pdata);
+	pinfo.fwnode = pdev->dev.fwnode; // copy fwnode for usb find of_node
 
 	glue->musb_pdev = platform_device_register_full(&pinfo);
 	if (IS_ERR(glue->musb_pdev)) {
