--- a/include/dt-bindings/clock/sun20i-d1-ccu.h
+++ b/include/dt-bindings/clock/sun20i-d1-ccu.h
@@ -79,78 +79,80 @@
 #define CLK_BUS_I2C1		69
 #define CLK_BUS_I2C2		70
 #define CLK_BUS_I2C3		71
-#define CLK_SPI0		72
-#define CLK_SPI1		73
-#define CLK_BUS_SPI0		74
-#define CLK_BUS_SPI1		75
-#define CLK_EMAC_25M		76
-#define CLK_BUS_EMAC		77
-#define CLK_IR_TX		78
-#define CLK_BUS_IR_TX		79
-#define CLK_BUS_GPADC		80
-#define CLK_BUS_THS		81
-#define CLK_I2S0		82
-#define CLK_I2S1		83
-#define CLK_I2S2		84
-#define CLK_I2S2_ASRC		85
-#define CLK_BUS_I2S0		86
-#define CLK_BUS_I2S1		87
-#define CLK_BUS_I2S2		88
-#define CLK_SPDIF_TX		89
-#define CLK_SPDIF_RX		90
-#define CLK_BUS_SPDIF		91
-#define CLK_DMIC		92
-#define CLK_BUS_DMIC		93
-#define CLK_AUDIO_DAC		94
-#define CLK_AUDIO_ADC		95
-#define CLK_BUS_AUDIO		96
-#define CLK_USB_OHCI0		97
-#define CLK_USB_OHCI1		98
-#define CLK_BUS_OHCI0		99
-#define CLK_BUS_OHCI1		100
-#define CLK_BUS_EHCI0		101
-#define CLK_BUS_EHCI1		102
-#define CLK_BUS_OTG		103
-#define CLK_BUS_LRADC		104
-#define CLK_BUS_DPSS_TOP	105
-#define CLK_HDMI_24M		106
-#define CLK_HDMI_CEC_32K	107
-#define CLK_HDMI_CEC		108
-#define CLK_BUS_HDMI		109
-#define CLK_MIPI_DSI		110
-#define CLK_BUS_MIPI_DSI	111
-#define CLK_TCON_LCD0		112
-#define CLK_BUS_TCON_LCD0	113
-#define CLK_TCON_TV		114
-#define CLK_BUS_TCON_TV		115
-#define CLK_TVE			116
-#define CLK_BUS_TVE_TOP		117
-#define CLK_BUS_TVE		118
-#define CLK_TVD			119
-#define CLK_BUS_TVD_TOP		120
-#define CLK_BUS_TVD		121
-#define CLK_LEDC		122
-#define CLK_BUS_LEDC		123
-#define CLK_CSI_TOP		124
-#define CLK_CSI_MCLK		125
-#define CLK_BUS_CSI		126
-#define CLK_TPADC		127
-#define CLK_BUS_TPADC		128
-#define CLK_BUS_TZMA		129
-#define CLK_DSP			130
-#define CLK_BUS_DSP_CFG		131
-#define CLK_RISCV		132
-#define CLK_RISCV_AXI		133
-#define CLK_BUS_RISCV_CFG	134
-#define CLK_FANOUT_24M		135
-#define CLK_FANOUT_12M		136
-#define CLK_FANOUT_16M		137
-#define CLK_FANOUT_25M		138
-#define CLK_FANOUT_32K		139
-#define CLK_FANOUT_27M		140
-#define CLK_FANOUT_PCLK		141
-#define CLK_FANOUT0		142
-#define CLK_FANOUT1		143
-#define CLK_FANOUT2		144
+#define CLK_BUS_CAN0		72
+#define CLK_BUS_CAN1		73
+#define CLK_SPI0		74
+#define CLK_SPI1		75
+#define CLK_BUS_SPI0		76
+#define CLK_BUS_SPI1		77
+#define CLK_EMAC_25M		78
+#define CLK_BUS_EMAC		79
+#define CLK_IR_TX		80
+#define CLK_BUS_IR_TX		81
+#define CLK_BUS_GPADC		82
+#define CLK_BUS_THS		83
+#define CLK_I2S0		84
+#define CLK_I2S1		85
+#define CLK_I2S2		86
+#define CLK_I2S2_ASRC		87
+#define CLK_BUS_I2S0		88
+#define CLK_BUS_I2S1		89
+#define CLK_BUS_I2S2		90
+#define CLK_SPDIF_TX		91
+#define CLK_SPDIF_RX		92
+#define CLK_BUS_SPDIF		93
+#define CLK_DMIC		94
+#define CLK_BUS_DMIC		95
+#define CLK_AUDIO_DAC		96
+#define CLK_AUDIO_ADC		97
+#define CLK_BUS_AUDIO		98
+#define CLK_USB_OHCI0		99
+#define CLK_USB_OHCI1		100
+#define CLK_BUS_OHCI0		101
+#define CLK_BUS_OHCI1		102
+#define CLK_BUS_EHCI0		103
+#define CLK_BUS_EHCI1		104
+#define CLK_BUS_OTG		105
+#define CLK_BUS_LRADC		106
+#define CLK_BUS_DPSS_TOP	107
+#define CLK_HDMI_24M		108
+#define CLK_HDMI_CEC_32K	109
+#define CLK_HDMI_CEC		110
+#define CLK_BUS_HDMI		111
+#define CLK_MIPI_DSI		112
+#define CLK_BUS_MIPI_DSI	113
+#define CLK_TCON_LCD0		114
+#define CLK_BUS_TCON_LCD0	115
+#define CLK_TCON_TV		116
+#define CLK_BUS_TCON_TV		117
+#define CLK_TVE			118
+#define CLK_BUS_TVE_TOP		119
+#define CLK_BUS_TVE		120
+#define CLK_TVD			121
+#define CLK_BUS_TVD_TOP		122
+#define CLK_BUS_TVD		123
+#define CLK_LEDC		124
+#define CLK_BUS_LEDC		125
+#define CLK_CSI_TOP		126
+#define CLK_CSI_MCLK		127
+#define CLK_BUS_CSI		128
+#define CLK_TPADC		129
+#define CLK_BUS_TPADC		130
+#define CLK_BUS_TZMA		131
+#define CLK_DSP			132
+#define CLK_BUS_DSP_CFG		133
+#define CLK_RISCV		134
+#define CLK_RISCV_AXI		135
+#define CLK_BUS_RISCV_CFG	136
+#define CLK_FANOUT_24M		137
+#define CLK_FANOUT_12M		138
+#define CLK_FANOUT_16M		139
+#define CLK_FANOUT_25M		140
+#define CLK_FANOUT_32K		141
+#define CLK_FANOUT_27M		142
+#define CLK_FANOUT_PCLK		143
+#define CLK_FANOUT0		144
+#define CLK_FANOUT1		145
+#define CLK_FANOUT2		146
 
 #endif /* _DT_BINDINGS_CLK_SUN20I_D1_CCU_H_ */
--- a/drivers/clk/sunxi-ng/ccu-sun20i-d1.c
+++ b/drivers/clk/sunxi-ng/ccu-sun20i-d1.c
@@ -469,6 +469,11 @@
 static SUNXI_CCU_GATE_HWS(bus_i2c3_clk, "bus-i2c3", apb1_hws,
 			  0x91c, BIT(3), 0);
 
+static SUNXI_CCU_GATE_HWS(bus_can0_clk, "bus-can0", apb1_hws,
+			  0x92c, BIT(0), 0);
+static SUNXI_CCU_GATE_HWS(bus_can1_clk, "bus-can1", apb1_hws,
+			  0x92c, BIT(1), 0);
+
 static const struct clk_parent_data spi_parents[] = {
 	{ .fw_name = "hosc" },
 	{ .hw = &pll_periph0_clk.hw },
@@ -1002,6 +1002,8 @@
 	&bus_i2c1_clk.common,
 	&bus_i2c2_clk.common,
 	&bus_i2c3_clk.common,
+	&bus_can0_clk.common,
+	&bus_can1_clk.common,
 	&spi0_clk.common,
 	&spi1_clk.common,
 	&bus_spi0_clk.common,
@@ -1147,6 +1154,8 @@
 		[CLK_BUS_I2C1]		= &bus_i2c1_clk.common.hw,
 		[CLK_BUS_I2C2]		= &bus_i2c2_clk.common.hw,
 		[CLK_BUS_I2C3]		= &bus_i2c3_clk.common.hw,
+		[CLK_BUS_CAN0]		= &bus_can0_clk.common.hw,
+		[CLK_BUS_CAN1]		= &bus_can1_clk.common.hw,
 		[CLK_SPI0]		= &spi0_clk.common.hw,
 		[CLK_SPI1]		= &spi1_clk.common.hw,
 		[CLK_BUS_SPI0]		= &bus_spi0_clk.common.hw,
@@ -1259,6 +1268,8 @@
 	[RST_BUS_I2C1]		= { 0x91c, BIT(17) },
 	[RST_BUS_I2C2]		= { 0x91c, BIT(18) },
 	[RST_BUS_I2C3]		= { 0x91c, BIT(19) },
+	[RST_BUS_CAN0]		= { 0x92c, BIT(16) },
+	[RST_BUS_CAN1]		= { 0x92c, BIT(17) },
 	[RST_BUS_SPI0]		= { 0x96c, BIT(16) },
 	[RST_BUS_SPI1]		= { 0x96c, BIT(17) },
 	[RST_BUS_EMAC]		= { 0x97c, BIT(16) },
--- a/include/dt-bindings/reset/sun20i-d1-ccu.h
+++ b/include/dt-bindings/reset/sun20i-d1-ccu.h
@@ -35,43 +35,45 @@
 #define RST_BUS_I2C1		25
 #define RST_BUS_I2C2		26
 #define RST_BUS_I2C3		27
-#define RST_BUS_SPI0		28
-#define RST_BUS_SPI1		29
-#define RST_BUS_EMAC		30
-#define RST_BUS_IR_TX		31
-#define RST_BUS_GPADC		32
-#define RST_BUS_THS		33
-#define RST_BUS_I2S0		34
-#define RST_BUS_I2S1		35
-#define RST_BUS_I2S2		36
-#define RST_BUS_SPDIF		37
-#define RST_BUS_DMIC		38
-#define RST_BUS_AUDIO		39
-#define RST_USB_PHY0		40
-#define RST_USB_PHY1		41
-#define RST_BUS_OHCI0		42
-#define RST_BUS_OHCI1		43
-#define RST_BUS_EHCI0		44
-#define RST_BUS_EHCI1		45
-#define RST_BUS_OTG		46
-#define RST_BUS_LRADC		47
-#define RST_BUS_DPSS_TOP	48
-#define RST_BUS_HDMI_SUB	49
-#define RST_BUS_HDMI_MAIN	50
-#define RST_BUS_MIPI_DSI	51
-#define RST_BUS_TCON_LCD0	52
-#define RST_BUS_TCON_TV		53
-#define RST_BUS_LVDS0		54
-#define RST_BUS_TVE		55
-#define RST_BUS_TVE_TOP		56
-#define RST_BUS_TVD		57
-#define RST_BUS_TVD_TOP		58
-#define RST_BUS_LEDC		59
-#define RST_BUS_CSI		60
-#define RST_BUS_TPADC		61
-#define RST_DSP			62
-#define RST_BUS_DSP_CFG		63
-#define RST_BUS_DSP_DBG		64
-#define RST_BUS_RISCV_CFG	65
+#define RST_BUS_CAN0		28
+#define RST_BUS_CAN1		29
+#define RST_BUS_SPI0		30
+#define RST_BUS_SPI1		31
+#define RST_BUS_EMAC		32
+#define RST_BUS_IR_TX		33
+#define RST_BUS_GPADC		34
+#define RST_BUS_THS		35
+#define RST_BUS_I2S0		36
+#define RST_BUS_I2S1		37
+#define RST_BUS_I2S2		38
+#define RST_BUS_SPDIF		39
+#define RST_BUS_DMIC		40
+#define RST_BUS_AUDIO		41
+#define RST_USB_PHY0		42
+#define RST_USB_PHY1		43
+#define RST_BUS_OHCI0		44
+#define RST_BUS_OHCI1		45
+#define RST_BUS_EHCI0		46
+#define RST_BUS_EHCI1		47
+#define RST_BUS_OTG		48
+#define RST_BUS_LRADC		49
+#define RST_BUS_DPSS_TOP	50
+#define RST_BUS_HDMI_SUB	51
+#define RST_BUS_HDMI_MAIN	52
+#define RST_BUS_MIPI_DSI	53
+#define RST_BUS_TCON_LCD0	54
+#define RST_BUS_TCON_TV		55
+#define RST_BUS_LVDS0		56
+#define RST_BUS_TVE		57
+#define RST_BUS_TVE_TOP		58
+#define RST_BUS_TVD		59
+#define RST_BUS_TVD_TOP		60
+#define RST_BUS_LEDC		61
+#define RST_BUS_CSI		62
+#define RST_BUS_TPADC		63
+#define RST_DSP			64
+#define RST_BUS_DSP_CFG		65
+#define RST_BUS_DSP_DBG		66
+#define RST_BUS_RISCV_CFG	67
 
 #endif /* _DT_BINDINGS_RST_SUN20I_D1_CCU_H_ */
