--- a/drivers/net/ethernet/stmicro/stmmac/stmmac.h
+++ b/drivers/net/ethernet/stmicro/stmmac/stmmac.h
@@ -324,6 +324,9 @@
 	/* XDP BPF Program */
 	unsigned long *af_xdp_zc_qps;
 	struct bpf_prog *xdp_prog;
+
+	/* For unregister_netdev crash */
+	int netdev_inited;
 };
 
 enum stmmac_state {
--- a/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
+++ b/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
@@ -7479,6 +7479,7 @@
 			__func__, ret);
 		goto error_netdev_register;
 	}
+	priv->netdev_inited = 1;
 
 #ifdef CONFIG_DEBUG_FS
 	stmmac_init_fs(ndev);
@@ -7529,8 +7530,10 @@
 
 	stmmac_stop_all_dma(priv);
 	stmmac_mac_set(priv, priv->ioaddr, false);
-	netif_carrier_off(ndev);
-	unregister_netdev(ndev);
+	if (priv->netdev_inited) {
+		netif_carrier_off(ndev);
+		unregister_netdev(ndev);
+	}
 
 #ifdef CONFIG_DEBUG_FS
 	stmmac_exit_fs(ndev);
