--- a/arch/arm/boot/dts/rockchip/rk3506.dtsi
+++ a/arch/arm/boot/dts/rockchip/rk3506.dtsi
@@ -192,45 +192,29 @@
 			     <GIC_PPI 11 (GIC_CPU_MASK_SIMPLE(3) | IRQ_TYPE_LEVEL_HIGH)>,
 			     <GIC_PPI 10 (GIC_CPU_MASK_SIMPLE(3) | IRQ_TYPE_LEVEL_HIGH)>;
 		clock-frequency = <24000000>;
+		always-on;
 	};
 
-	uart0: serial@ff0a0000 {
-		compatible = "rockchip,rk3506-uart", "snps,dw-apb-uart";
-		reg = <0xff0a0000 0x100>;
-		interrupts = <GIC_SPI 34 IRQ_TYPE_LEVEL_HIGH>;
-		reg-shift = <2>;
-		reg-io-width = <4>;
-		clocks = <&cru SCLK_UART0>, <&cru PCLK_UART0>;
-		clock-names = "baudclk", "apb_pclk";
-		pinctrl-names = "default";
-		pinctrl-0 = <&uart0_xfer_pins>;
-		status = "disabled";
-	};
-	uart1: serial@ff0b0000 {
-		compatible = "rockchip,rk3506-uart", "snps,dw-apb-uart";
-		reg = <0xff0b0000 0x100>;
-		interrupts = <GIC_SPI 35 IRQ_TYPE_LEVEL_HIGH>;
-		reg-shift = <2>;
-		reg-io-width = <4>;
-		clocks = <&cru SCLK_UART1>, <&cru PCLK_UART1>;
-		clock-names = "baudclk", "apb_pclk";
-		pinctrl-names = "default";
-		status = "disabled";
+	dmac0: dma-controller@ff000000 {
+		compatible = "arm,pl330", "arm,primecell";
+		reg = <0xff000000 0x4000>;
+		interrupts = <GIC_SPI 116 IRQ_TYPE_LEVEL_HIGH>,
+			     <GIC_SPI 117 IRQ_TYPE_LEVEL_HIGH>;
+		clocks = <&cru ACLK_DMAC0>;
+		clock-names = "apb_pclk";
+		#dma-cells = <5>;
+		arm,pl330-periph-burst;
 	};
 
-
-	spi0: spi@ff120000 {
-		compatible = "rockchip,rk3506-spi", "rockchip,rk3066-spi";
-		reg = <0xff120000 0x1000>;
-		interrupts = <GIC_SPI 43 IRQ_TYPE_LEVEL_HIGH>;
-		#address-cells = <1>;
-		#size-cells = <0>;
-		clocks = <&cru CLK_SPI0>, <&cru PCLK_SPI0>;
-		clock-names = "spiclk", "apb_pclk";
-		num-cs = <2>;
-		pinctrl-names = "default";
-		pinctrl-0 = <&spi0_csn0_pins &spi0_csn1_pins &spi0_clk_pins>;
-		status = "disabled";
+	dmac1: dma-controller@ff008000 {
+		compatible = "arm,pl330", "arm,primecell";
+		reg = <0xff008000 0x4000>;
+		interrupts = <GIC_SPI 118 IRQ_TYPE_LEVEL_HIGH>,
+			     <GIC_SPI 119 IRQ_TYPE_LEVEL_HIGH>;
+		clocks = <&cru ACLK_DMAC1>;
+		clock-names = "apb_pclk";
+		#dma-cells = <5>;
+		arm,pl330-periph-burst;
 	};
 
 	i2c0: i2c@ff040000 {
@@ -266,12 +250,44 @@
 		status = "disabled";
 	};
 
+
+	uart0: serial@ff0a0000 {
+		compatible = "rockchip,rk3506-uart", "snps,dw-apb-uart";
+		reg = <0xff0a0000 0x100>;
+		interrupts = <GIC_SPI 34 IRQ_TYPE_LEVEL_HIGH>;
+		reg-shift = <2>;
+		reg-io-width = <4>;
+		dmas = <&dmac0 4 0xff2880a8 0x03000100 0x0 0x0>,
+		       <&dmac0 5 0xff2880a8 0x0c000400 0x0 0x0>;
+		clocks = <&cru SCLK_UART0>, <&cru PCLK_UART0>;
+		clock-names = "baudclk", "apb_pclk";
+		pinctrl-names = "default";
+		pinctrl-0 = <&uart0_xfer_pins>;
+		status = "disabled";
+	};
+
+	uart1: serial@ff0b0000 {
+		compatible = "rockchip,rk3506-uart", "snps,dw-apb-uart";
+		reg = <0xff0b0000 0x100>;
+		interrupts = <GIC_SPI 35 IRQ_TYPE_LEVEL_HIGH>;
+		reg-shift = <2>;
+		reg-io-width = <4>;
+		dmas = <&dmac0 6 0xff2880a8 0x30001000 0x0 0x0>,
+		       <&dmac0 7 0xff2880a8 0xc0004000 0x0 0x0>;
+		clocks = <&cru SCLK_UART1>, <&cru PCLK_UART1>;
+		clock-names = "baudclk", "apb_pclk";
+		pinctrl-names = "default";
+		status = "disabled";
+	};
+
 	uart2: serial@ff0c0000 {
 		compatible = "rockchip,rk3506-uart", "snps,dw-apb-uart";
 		reg = <0xff0c0000 0x100>;
 		interrupts = <GIC_SPI 36 IRQ_TYPE_LEVEL_HIGH>;
 		reg-shift = <2>;
 		reg-io-width = <4>;
+		dmas = <&dmac0 8 0xff2880ac 0x00030001 0x0 0x0>,
+		       <&dmac0 9 0xff2880ac 0x000c0004 0x0 0x0>;
 		clocks = <&cru SCLK_UART2>, <&cru PCLK_UART2>;
 		clock-names = "baudclk", "apb_pclk";
 		status = "disabled";
@@ -283,32 +299,58 @@
 		interrupts = <GIC_SPI 37 IRQ_TYPE_LEVEL_HIGH>;
 		reg-shift = <2>;
 		reg-io-width = <4>;
+		dmas = <&dmac0 10 0xff2880ac 0x00300010 0x0 0x0>,
+		       <&dmac0 11 0xff2880ac 0x00c00040 0x0 0x0>;
 		clocks = <&cru SCLK_UART3>, <&cru PCLK_UART3>;
 		clock-names = "baudclk", "apb_pclk";
 		pinctrl-names = "default";
 		status = "disabled";
 	};
 
-	dmac0: dma-controller@ff000000 {
-		compatible = "arm,pl330", "arm,primecell";
-		reg = <0xff000000 0x4000>;
-		interrupts = <GIC_SPI 116 IRQ_TYPE_LEVEL_HIGH>,
-			     <GIC_SPI 117 IRQ_TYPE_LEVEL_HIGH>;
-		clocks = <&cru ACLK_DMAC0>;
-		clock-names = "apb_pclk";
-		#dma-cells = <5>;
-		arm,pl330-periph-burst;
+	spi0: spi@ff120000 {
+		compatible = "rockchip,rk3506-spi", "rockchip,rk3066-spi";
+		reg = <0xff120000 0x1000>;
+		interrupts = <GIC_SPI 43 IRQ_TYPE_LEVEL_HIGH>;
+		#address-cells = <1>;
+		#size-cells = <0>;
+		clocks = <&cru CLK_SPI0>, <&cru PCLK_SPI0>;
+		clock-names = "spiclk", "apb_pclk";
+		num-cs = <2>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&spi0_csn0_pins &spi0_csn1_pins &spi0_clk_pins>;
+		status = "disabled";
 	};
 
-	dmac1: dma-controller@ff008000 {
-		compatible = "arm,pl330", "arm,primecell";
-		reg = <0xff008000 0x4000>;
-		interrupts = <GIC_SPI 118 IRQ_TYPE_LEVEL_HIGH>,
-			     <GIC_SPI 119 IRQ_TYPE_LEVEL_HIGH>;
-		clocks = <&cru ACLK_DMAC1>;
-		clock-names = "apb_pclk";
-		#dma-cells = <5>;
-		arm,pl330-periph-burst;
+	grf: syscon@ff288000 {
+		compatible = "rockchip,rk3506-grf", "syscon", "simple-mfd";
+		reg = <0xff288000 0x4000>;
+	};
+
+	usb2phy: usb2-phy@ff2b0000 {
+		compatible = "rockchip,rk3506-usb2phy";
+		reg = <0xff2b0000 0x8000>;
+		clocks = <&cru CLK_REF_USBPHY_TOP>, <&cru PCLK_USBPHY>;
+		clock-names = "phyclk", "apb_pclk";
+		rockchip,usbgrf = <&grf>;
+		status = "disabled";
+
+		u2phy_otg0: otg-port {
+			#phy-cells = <0>;
+			interrupts = <GIC_SPI 75 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 76 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 77 IRQ_TYPE_LEVEL_HIGH>;
+			interrupt-names = "otg-bvalid", "otg-id", "linestate";
+			status = "disabled";
+		};
+
+		u2phy_otg1: host-port {
+			#phy-cells = <0>;
+			interrupts = <GIC_SPI 80 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 81 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 82 IRQ_TYPE_LEVEL_HIGH>;
+			interrupt-names = "otg-bvalid", "otg-id", "linestate";
+			status = "disabled";
+		};
 	};
 
 	sai1: sai@ff310000 {
@@ -361,8 +403,8 @@
 		rockchip,max-dll = <0x17F>;	/* RK3506 SFC is VER_5; cap DLL cell scan (vendor value) */
 		#address-cells = <1>;
 		#size-cells = <0>;
-		// pinctrl-names = "default"; // The dedfault pin setting is ready to work, do not set this, or DLL delay tuning will be failed.
-		// pinctrl-0 = <&fspi_bus4_pins &fspi_clk_pins &fspi_csn_pins>; 
+		pinctrl-names = "default"; // The dedfault pin setting is ready to work, do not set this, or DLL delay tuning will be failed.
+		pinctrl-0 = <&fspi_bus4_pins &fspi_clk_pins &fspi_csn_pins>; 
 		status = "disabled";
 	};
 
@@ -474,38 +516,6 @@
 		};
 	};
 
-	grf: syscon@ff288000 {
-		compatible = "rockchip,rk3506-grf", "syscon", "simple-mfd";
-		reg = <0xff288000 0x4000>;
-	};
-
-	usb2phy: usb2-phy@ff2b0000 {
-		compatible = "rockchip,rk3506-usb2phy";
-		reg = <0xff2b0000 0x8000>;
-		clocks = <&cru CLK_REF_USBPHY_TOP>, <&cru PCLK_USBPHY>;
-		clock-names = "phyclk", "apb_pclk";
-		rockchip,usbgrf = <&grf>;
-		status = "disabled";
-
-		u2phy_otg0: otg-port {
-			#phy-cells = <0>;
-			interrupts = <GIC_SPI 75 IRQ_TYPE_LEVEL_HIGH>,
-				     <GIC_SPI 76 IRQ_TYPE_LEVEL_HIGH>,
-				     <GIC_SPI 77 IRQ_TYPE_LEVEL_HIGH>;
-			interrupt-names = "otg-bvalid", "otg-id", "linestate";
-			status = "disabled";
-		};
-
-		u2phy_otg1: host-port {
-			#phy-cells = <0>;
-			interrupts = <GIC_SPI 80 IRQ_TYPE_LEVEL_HIGH>,
-				     <GIC_SPI 81 IRQ_TYPE_LEVEL_HIGH>,
-				     <GIC_SPI 82 IRQ_TYPE_LEVEL_HIGH>;
-			interrupt-names = "otg-bvalid", "otg-id", "linestate";
-			status = "disabled";
-		};
-	};
-
 	ioc_grf: syscon@ff4d8000 {
 		compatible = "rockchip,rk3506-ioc-grf", "syscon";
 		reg = <0xff4d8000 0x8000>;
@@ -684,6 +694,11 @@
 			bias-pull-up;
 		};
 
+		pcfg_pull_up_drv_level_2: pcfg-pull-up-drv-level-2 {
+			bias-pull-up;
+			drive-strength = <2>;
+		};
+
 		pcfg_pull_none: pcfg-pull-none {
 			bias-disable;
 		};
@@ -693,6 +708,15 @@
 			drive-strength = <3>;
 		};
 
+		pcfg_pull_down: pcfg-pull-down {
+			bias-pull-down;
+		};
+
+		pcfg_pull_down_drv_level_3: pcfg-pull-down-drv-level-3 {
+			bias-pull-down;
+			drive-strength = <3>;
+		};
+
 		gpio0: gpio@ff940000 {
 			compatible = "rockchip,gpio-bank";
 			reg = <0xff940000 0x200>;
@@ -965,26 +989,26 @@
 			fspi_bus4_pins: fspi-bus4-pins {
 				rockchip,pins =
 					/* fspi_d0 */
-					<2 RK_PA2 1 &pcfg_pull_none>,
+					<2 RK_PA2 1 &pcfg_pull_up_drv_level_2>,
 					/* fspi_d1 */
-					<2 RK_PA3 1 &pcfg_pull_none>,
+					<2 RK_PA3 1 &pcfg_pull_up_drv_level_2>,
 					/* fspi_d2 */
-					<2 RK_PA4 1 &pcfg_pull_none>,
+					<2 RK_PA4 1 &pcfg_pull_up_drv_level_2>,
 					/* fspi_d3 */
-					<2 RK_PA5 1 &pcfg_pull_none>;
+					<2 RK_PA5 1 &pcfg_pull_up_drv_level_2>;
 			};
 
 			/omit-if-no-ref/
 			fspi_clk_pins: fspi-clk-pins {
 				rockchip,pins =
 					/* fspi_clk */
-					<2 RK_PA1 1 &pcfg_pull_none>;
+					<2 RK_PA1 1 &pcfg_pull_down_drv_level_3>;
 			};
 			/omit-if-no-ref/
 			fspi_csn_pins: fspi-csn-pins {
 				rockchip,pins =
 					/* fspi_csn */
-					<2 RK_PA0 1 &pcfg_pull_none>;
+					<2 RK_PA0 1 &pcfg_pull_up_drv_level_2>;
 			};
 		};
 	};
