--- a/drivers/net/phy/realtek/realtek_main.c
+++ b/drivers/net/phy/realtek/realtek_main.c
@@ -406,6 +406,19 @@ static int rtl8211f_config_init(struct p
 		return ret;
 	}

+	/*
+	ret = phy_read_paged(phydev, 0xd04, 0x10);
+	dev_info(dev, "RTL8211F_LCR:%x\n", ret);
+	ret = phy_read_paged(phydev, 0xd04, 0x11);
+	dev_info(dev, "RTL8211F_EEELCR:%x\n", ret);
+	*/
+	// Disable EEE led status (blink only transfer data)
+	ret = phy_modify_paged_changed(phydev, 0xd04, 0x11, 0xe, 0);
+	/*
+	ret = phy_read_paged(phydev, 0xd04, 0x11);
+	dev_info(dev, "RTL8211F_EEELCR:%x\n", ret);
+	*/
+
 	switch (phydev->interface) {
 	case PHY_INTERFACE_MODE_RGMII:
 		val_txdly = 0;
