--- a/drivers/usb/serial/option.c	2021-12-09 13:27:51.486959755 +0800
+++ b/drivers/usb/serial/option.c	2021-12-09 13:33:28.643788446 +0800
@@ -565,6 +565,12 @@
 #define WETELECOM_PRODUCT_6802			0x6802
 #define WETELECOM_PRODUCT_WMD300		0x6803
 
+/* Fibocom products */
+#define FIBOCOM_VENDOR_ID			0x1782
+#define FIBOCOM_PRODUCT_MC61531		0x4d10
+#define FIBOCOM_PRODUCT_MC61532		0x4d11
+#define FIBOCOM_PRODUCT_MC61533		0x4d12
+
 
 /* Device flags */
 
@@ -2099,6 +2105,9 @@
 	  .driver_info = RSVD(6) },
 	{ USB_DEVICE_AND_INTERFACE_INFO(0x2cb7, 0x010b, 0xff, 0xff, 0x30) },	/* Fibocom FG150 Diag */
 	{ USB_DEVICE_AND_INTERFACE_INFO(0x2cb7, 0x010b, 0xff, 0, 0) },		/* Fibocom FG150 AT */
+	{ USB_DEVICE(FIBOCOM_VENDOR_ID, FIBOCOM_PRODUCT_MC61531) },		/* Fibocom MC61531 */
+	{ USB_DEVICE(FIBOCOM_VENDOR_ID, FIBOCOM_PRODUCT_MC61532) },		/* Fibocom MC61532 */
+	{ USB_DEVICE(FIBOCOM_VENDOR_ID, FIBOCOM_PRODUCT_MC61533) },		/* Fibocom MC61533 */
 	{ USB_DEVICE_INTERFACE_CLASS(0x2cb7, 0x01a0, 0xff) },			/* Fibocom NL668-AM/NL652-EU (laptop MBIM) */
 	{ USB_DEVICE_INTERFACE_CLASS(0x2cb7, 0x01a2, 0xff) },			/* Fibocom FM101-GL (laptop MBIM) */
 	{ USB_DEVICE_INTERFACE_CLASS(0x2cb7, 0x01a4, 0xff),			/* Fibocom FM101-GL (laptop MBIM) */
