--- a/drivers/net/ethernet/stmicro/stmmac/dwmac-sun8i.c
+++ b/drivers/net/ethernet/stmicro/stmmac/dwmac-sun8i.c
@@ -1045,8 +1045,14 @@
 	if (gmac->variant->soc_has_internal_phy)
 		sun8i_dwmac_unpower_internal_phy(gmac);
 
-	regulator_disable(gmac->regulator_phy);
-	regulator_disable(gmac->regulator_phy_io);
+	if (gmac->regulator_phy) {
+		regulator_disable(gmac->regulator_phy);
+		gmac->regulator_phy = NULL;
+	}
+	if (gmac->regulator_phy_io) {
+		regulator_disable(gmac->regulator_phy_io);
+		gmac->regulator_phy_io = NULL;
+	}
 }
 
 static void sun8i_dwmac_set_mac_loopback(void __iomem *ioaddr, bool enable)
